{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "825b615a-d04a-4c3e-8a14-a80257e2ed18",
   "metadata": {},
   "source": [
    "# Clinical, Genomic, and Imaging data - Training and Testing "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "9515c767-f825-4691-a8dc-ebe2016e6ae2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import boto3\n",
    "import sagemaker\n",
    "from sagemaker.session import Session\n",
    "from sagemaker import get_execution_role\n",
    "from sagemaker.serializers import CSVSerializer\n",
    "from sagemaker.image_uris import retrieve\n",
    "from sagemaker.session import TrainingInput\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.io as sio\n",
    "import io, os\n",
    "import sys\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "import matplotlib.gridspec as gridspec\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc14ce54-3d26-4810-a708-25574533e8c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install autogluon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f54c292-5f3c-48d8-b75c-ff8a7b6d6935",
   "metadata": {},
   "outputs": [],
   "source": [
    "from autogluon.tabular import TabularPredictor\n",
    "import autogluon as ag"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aaa341b3-3db4-42ca-b40d-ed3f904af38e",
   "metadata": {},
   "source": [
    "## Get data type to train model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "9588fcd1-5671-4266-bc97-8607c4d6b6ca",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "supported_data_type = ('genomic', 'genomic-clinical', 'genomic-clinical-imaging', 'clinical', 'imaging')\n",
    "#data_type = 'genomic'\n",
    "#data_type = 'genomic-clinical'\n",
    "data_type = 'genomic-clinical-imaging'\n",
    "# data_type = 'clinical'\n",
    "#data_type = 'imaging'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f648bda3-d7c7-4c0e-984a-d96e925e25fd",
   "metadata": {},
   "source": [
    "## Set up S3 buckets and session"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "191ef27d-efbe-4229-a2c8-2f4fcd02b581",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sm_session = sagemaker.Session()\n",
    "bucket = sm_session.default_bucket()\n",
    "region = boto3.Session().region_name\n",
    "role = get_execution_role()\n",
    "\n",
    "boto_session = boto3.Session(region_name=region)\n",
    "sagemaker_client = boto_session.client(service_name='sagemaker', region_name=region)\n",
    "featurestore_runtime = boto_session.client(service_name='sagemaker-featurestore-runtime', region_name=region)\n",
    "\n",
    "feature_store_session = Session(\n",
    "    boto_session=boto_session,\n",
    "    sagemaker_client=sagemaker_client,\n",
    "    sagemaker_featurestore_runtime_client=featurestore_runtime\n",
    ")\n",
    "\n",
    "s3_client = boto3.client('s3', region_name=region)\n",
    "\n",
    "default_s3_bucket_name = sm_session.default_bucket()\n",
    "prefix = 'multi-model-health-ml'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "043669dd-325a-46ac-a5ee-21cca32a9470",
   "metadata": {},
   "source": [
    "## Get features from SageMaker FeatureStore based on data type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0e6a0bbe-aae3-4fdd-bca0-7507c935012e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from sagemaker.feature_store.feature_group import FeatureGroup\n",
    "\n",
    "genomic_feature_group_name = 'genomic-feature-group'\n",
    "clinical_feature_group_name = 'clinical-feature-group'\n",
    "imaging_feature_group_name = 'imaging-feature-group'\n",
    "\n",
    "genomic_feature_group = FeatureGroup(name=genomic_feature_group_name, sagemaker_session=feature_store_session)\n",
    "clinical_feature_group = FeatureGroup(name=clinical_feature_group_name, sagemaker_session=feature_store_session)\n",
    "imaging_feature_group = FeatureGroup(name=imaging_feature_group_name, sagemaker_session=feature_store_session)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "eef950fe-e6cd-46f7-ab9a-ba2416af69be",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table names\n",
      "genomic_feature_group_1688073848\n",
      "clinical_feature_group_1688073736\n",
      "imaging_feature_group_1688074260\n"
     ]
    }
   ],
   "source": [
    "genomic_query = genomic_feature_group.athena_query()\n",
    "clinical_query = clinical_feature_group.athena_query()\n",
    "imaging_query = imaging_feature_group.athena_query()\n",
    "\n",
    "genomic_table = genomic_query.table_name\n",
    "clinical_table = clinical_query.table_name\n",
    "imaging_table = imaging_query.table_name\n",
    "\n",
    "print('Table names')\n",
    "print(genomic_table)\n",
    "print(clinical_table)\n",
    "print(imaging_table)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "c25be564-ebb7-429e-936e-6e00e4fd89d2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_features(data_type, output_location):  \n",
    "    if (data_type == 'clinical'):\n",
    "        query_string = f'SELECT * FROM \"{clinical_table}\" ORDER BY patientid'\n",
    "        print(query_string)\n",
    "\n",
    "        clinical_query.run(query_string=query_string, output_location=output_location)\n",
    "        clinical_query.wait()\n",
    "        dataset = clinical_query.as_dataframe()\n",
    "        \n",
    "    elif (data_type == 'imaging'):\n",
    "        query_string = f'SELECT * FROM \"{imaging_table}\" ORDER BY patientid'\n",
    "        print(query_string)\n",
    "\n",
    "        imaging_query.run(query_string=query_string, output_location=output_location)\n",
    "        imaging_query.wait()\n",
    "        dataset = imaging_query.as_dataframe()\n",
    "    \n",
    "    elif (data_type == 'genomic'):\n",
    "        query_string = f'SELECT * FROM \"{genomic_table}\" ORDER BY patientid'\n",
    "        print(query_string)\n",
    "        genomic_query.run(query_string=query_string, output_location=output_location)\n",
    "        genomic_query.wait()\n",
    "        dataset = genomic_query.as_dataframe()\n",
    "        \n",
    "    elif (data_type == 'genomic-clinical'):\n",
    "        query_string = f'''SELECT * FROM \"{clinical_table}\", \"{genomic_table}\"\n",
    "                           WHERE \"{clinical_table}\".patientid = \"{genomic_table}\".patientid\n",
    "                           ORDER BY \"{clinical_table}\".patientid ASC'''\n",
    "        print(query_string)\n",
    "\n",
    "        genomic_query.run(query_string=query_string, output_location=output_location)\n",
    "        genomic_query.wait()\n",
    "        dataset = genomic_query.as_dataframe()\n",
    "        \n",
    "    elif (data_type == 'genomic-clinical-imaging'):\n",
    "        query_string = f'''SELECT * FROM \"{genomic_table}\", \"{clinical_table}\", \"{imaging_table}\"\n",
    "                           WHERE \"{genomic_table}\".patientid = \"{clinical_table}\".patientid\n",
    "                           AND \"{genomic_table}\".patientid = \"{imaging_table}\".patientid\n",
    "                           ORDER BY \"{clinical_table}\".patientid ASC'''                   \n",
    "        print(query_string)\n",
    "        \n",
    "        genomic_query.run(query_string=query_string, output_location=output_location)\n",
    "        genomic_query.wait()\n",
    "        dataset = genomic_query.as_dataframe()\n",
    "        \n",
    "    elif data_type not in supported_data_type:\n",
    "        raise KeyError(f'data_type {data_type} is not supported for this analysis.')\n",
    "        \n",
    "    return dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "7e6f4687-3bd3-4c1d-8db7-9d164a8841a3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SELECT * FROM \"genomic_feature_group_1688073848\", \"clinical_feature_group_1688073736\", \"imaging_feature_group_1688074260\"\n",
      "                           WHERE \"genomic_feature_group_1688073848\".patientid = \"clinical_feature_group_1688073736\".patientid\n",
      "                           AND \"genomic_feature_group_1688073848\".patientid = \"imaging_feature_group_1688074260\".patientid\n",
      "                           ORDER BY \"clinical_feature_group_1688073736\".patientid ASC\n"
     ]
    }
   ],
   "source": [
    "fs_output_location = f's3://{default_s3_bucket_name}/{prefix}/feature-store-queries'\n",
    "dataset = get_features(data_type, fs_output_location)\n",
    "dataset = dataset.astype(str).replace({\"{\":\"\", \"}\":\"\"}, regex=True)\n",
    "\n",
    "# Write to csv in S3 without headers and index column.\n",
    "filename=f'{data_type}-dataset.csv'\n",
    "dataset_uri_prefix = f's3://{default_s3_bucket_name}/{prefix}/training_input/';\n",
    "\n",
    "dataset.to_csv(filename)\n",
    "s3_client.upload_file(filename, default_s3_bucket_name, f'{prefix}/training_input/{filename}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "6aea7d6a-1b2e-4019-9866-a2a27b95ecb9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patientid</th>\n",
       "      <th>gene_info</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>contigname</th>\n",
       "      <th>start</th>\n",
       "      <th>referenceallele</th>\n",
       "      <th>alternatealleles</th>\n",
       "      <th>phased</th>\n",
       "      <th>calls</th>\n",
       "      <th>alzheimers_prediction</th>\n",
       "      <th>...</th>\n",
       "      <th>original_ngtdm_strength</th>\n",
       "      <th>imagesetid</th>\n",
       "      <th>alzheimers_prediction.2</th>\n",
       "      <th>coronary_heart_disease_prediction.2</th>\n",
       "      <th>stroke_prediction.2</th>\n",
       "      <th>hypertension_prediction.2</th>\n",
       "      <th>eventtime.2</th>\n",
       "      <th>write_time.2</th>\n",
       "      <th>api_invocation_time.2</th>\n",
       "      <th>is_deleted.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0074596f-5fd0-7965-db0f-cce71c81567d</td>\n",
       "      <td>'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...</td>\n",
       "      <td>'Conflicting_interpretations_of_pathogenicity'...</td>\n",
       "      <td>'14', '9', '6', '16', '17', '8', '10', '11', '...</td>\n",
       "      <td>53767041, 44919688, 31667848, 12351625, 449086...</td>\n",
       "      <td>'T', 'G', 'A', 'C'</td>\n",
       "      <td>'[A]', '[C]', '[G]', '[T]'</td>\n",
       "      <td>False</td>\n",
       "      <td>'[1, 1]', '[0, 1]'</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.547276310306838</td>\n",
       "      <td>'d7298fe7dde4537b8343b5a702979aea'</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1688074260.0</td>\n",
       "      <td>2023-06-29 21:37:00.939</td>\n",
       "      <td>2023-06-29 21:31:33.000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0618424e-ed51-3100-ea5c-e46492bfd65b</td>\n",
       "      <td>'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...</td>\n",
       "      <td>'Conflicting_interpretations_of_pathogenicity'...</td>\n",
       "      <td>'14', '9', '6', '16', '17', '8', '10', '11', '...</td>\n",
       "      <td>53767041, 44919688, 12351625, 31667848, 449086...</td>\n",
       "      <td>'T', 'G', 'A', 'C'</td>\n",
       "      <td>'[A]', '[C]', '[G]', '[T]'</td>\n",
       "      <td>False</td>\n",
       "      <td>'[1, 1]', '[0, 1]'</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>9.241749116927542</td>\n",
       "      <td>'e5e2ccf0487eed5523395431675fa708'</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1688074260.0</td>\n",
       "      <td>2023-06-29 21:36:51.705</td>\n",
       "      <td>2023-06-29 21:31:33.000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>06cc033a-f09a-0fb2-4a1a-4c4d99d88839</td>\n",
       "      <td>'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...</td>\n",
       "      <td>'Conflicting_interpretations_of_pathogenicity'...</td>\n",
       "      <td>'14', '9', '6', '16', '17', '8', '10', '11', '...</td>\n",
       "      <td>53767041, 44919688, 31667848, 12351625, 449086...</td>\n",
       "      <td>'T', 'G', 'A', 'C'</td>\n",
       "      <td>'[A]', '[C]', '[G]', '[T]'</td>\n",
       "      <td>False</td>\n",
       "      <td>'[1, 1]', '[0, 1]'</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>9.02168678253902</td>\n",
       "      <td>'1a3c6b0044e2b58a13c106896da369ef'</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1688074260.0</td>\n",
       "      <td>2023-06-29 21:37:00.886</td>\n",
       "      <td>2023-06-29 21:31:33.000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>06d43bcb-8322-3c2b-40f1-189b2852b5ab</td>\n",
       "      <td>'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...</td>\n",
       "      <td>'Conflicting_interpretations_of_pathogenicity'...</td>\n",
       "      <td>'14', '9', '6', '16', '17', '8', '10', '11', '...</td>\n",
       "      <td>53767041, 44919688, 12351625, 31667848, 449086...</td>\n",
       "      <td>'T', 'G', 'A', 'C'</td>\n",
       "      <td>'[A]', '[C]', '[G]', '[T]'</td>\n",
       "      <td>False</td>\n",
       "      <td>'[1, 1]', '[0, 1]'</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.776968231552623</td>\n",
       "      <td>'41b14124cdd7f0ea3d459bcfe6120e7f'</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1688074260.0</td>\n",
       "      <td>2023-06-29 21:36:51.674</td>\n",
       "      <td>2023-06-29 21:31:33.000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>08acefb1-271e-01d2-1bb6-aae9f65dbb42</td>\n",
       "      <td>'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...</td>\n",
       "      <td>'Conflicting_interpretations_of_pathogenicity'...</td>\n",
       "      <td>'14', '9', '6', '16', '17', '8', '10', '11', '...</td>\n",
       "      <td>53767041, 44919688, 12351625, 31667848, 449086...</td>\n",
       "      <td>'T', 'G', 'A', 'C'</td>\n",
       "      <td>'[A]', '[C]', '[G]', '[T]'</td>\n",
       "      <td>False</td>\n",
       "      <td>'[1, 1]', '[0, 1]'</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.601098896888011</td>\n",
       "      <td>'dd9a6771bf229a91f03f608c336141de'</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1688074260.0</td>\n",
       "      <td>2023-06-29 21:36:59.693</td>\n",
       "      <td>2023-06-29 21:31:33.000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0b73e9c9-18c0-5c0e-c5d5-95d7ca56c050</td>\n",
       "      <td>'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...</td>\n",
       "      <td>'Conflicting_interpretations_of_pathogenicity'...</td>\n",
       "      <td>'14', '9', '6', '16', '17', '8', '10', '11', '...</td>\n",
       "      <td>53767041, 44919688, 31667848, 12351625, 449086...</td>\n",
       "      <td>'T', 'G', 'A', 'C'</td>\n",
       "      <td>'[A]', '[C]', '[G]', '[T]'</td>\n",
       "      <td>False</td>\n",
       "      <td>'[1, 1]', '[0, 1]'</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>9.106481386718237</td>\n",
       "      <td>'1526fea901654ad5e45f7d5f22377f81'</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1688074260.0</td>\n",
       "      <td>2023-06-29 21:36:51.701</td>\n",
       "      <td>2023-06-29 21:31:33.000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0bb4b1f7-4128-89ef-a023-b81379c30f4f</td>\n",
       "      <td>'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...</td>\n",
       "      <td>'Conflicting_interpretations_of_pathogenicity'...</td>\n",
       "      <td>'14', '9', '6', '16', '17', '8', '10', '11', '...</td>\n",
       "      <td>53767041, 44919688, 31667848, 12351625, 449086...</td>\n",
       "      <td>'T', 'G', 'A', 'C'</td>\n",
       "      <td>'[A]', '[C]', '[G]', '[T]'</td>\n",
       "      <td>False</td>\n",
       "      <td>'[1, 1]', '[0, 1]'</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>9.38894545484775</td>\n",
       "      <td>'b94c9e7c782da3660ea6a905e2a91b34'</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1688074260.0</td>\n",
       "      <td>2023-06-29 21:37:00.006</td>\n",
       "      <td>2023-06-29 21:31:33.000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0c8ad7c0-b403-98d1-d6c9-ff1dbbce7d7c</td>\n",
       "      <td>'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...</td>\n",
       "      <td>'Conflicting_interpretations_of_pathogenicity'...</td>\n",
       "      <td>'14', '9', '6', '16', '17', '8', '10', '11', '...</td>\n",
       "      <td>53767041, 44919688, 31667848, 12351625, 449086...</td>\n",
       "      <td>'T', 'G', 'A', 'C'</td>\n",
       "      <td>'[A]', '[C]', '[G]', '[T]'</td>\n",
       "      <td>False</td>\n",
       "      <td>'[1, 1]', '[0, 1]'</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>9.202929567908852</td>\n",
       "      <td>'bccc8f07e6b97098dbb92cbafef2949c'</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1688074260.0</td>\n",
       "      <td>2023-06-29 21:36:59.699</td>\n",
       "      <td>2023-06-29 21:31:33.000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0f1c1042-f103-e85b-1a7a-03442861947d</td>\n",
       "      <td>'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...</td>\n",
       "      <td>'Conflicting_interpretations_of_pathogenicity'...</td>\n",
       "      <td>'14', '9', '6', '16', '17', '8', '10', '11', '...</td>\n",
       "      <td>53767041, 44919688, 12351625, 31667848, 449086...</td>\n",
       "      <td>'T', 'G', 'A', 'C'</td>\n",
       "      <td>'[A]', '[C]', '[G]', '[T]'</td>\n",
       "      <td>False</td>\n",
       "      <td>'[1, 1]', '[0, 1]'</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>11.269159030234578</td>\n",
       "      <td>'df83b4160e70a45bee0af99f9785d1eb'</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1688074260.0</td>\n",
       "      <td>2023-06-29 21:36:59.693</td>\n",
       "      <td>2023-06-29 21:31:33.000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>14e884a7-dc93-7909-9585-2857c6eb29e2</td>\n",
       "      <td>'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...</td>\n",
       "      <td>'Conflicting_interpretations_of_pathogenicity'...</td>\n",
       "      <td>'14', '9', '6', '16', '17', '8', '10', '11', '...</td>\n",
       "      <td>53767041, 44919688, 12351625, 31667848, 449086...</td>\n",
       "      <td>'T', 'G', 'A', 'C'</td>\n",
       "      <td>'[A]', '[C]', '[G]', '[T]'</td>\n",
       "      <td>False</td>\n",
       "      <td>'[1, 1]', '[0, 1]'</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.875513496551257</td>\n",
       "      <td>'20e77ca11e2b601b8ff9b9743efc1014'</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1688074260.0</td>\n",
       "      <td>2023-06-29 21:37:00.854</td>\n",
       "      <td>2023-06-29 21:31:33.000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 171 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              patientid  \\\n",
       "0  0074596f-5fd0-7965-db0f-cce71c81567d   \n",
       "1  0618424e-ed51-3100-ea5c-e46492bfd65b   \n",
       "2  06cc033a-f09a-0fb2-4a1a-4c4d99d88839   \n",
       "3  06d43bcb-8322-3c2b-40f1-189b2852b5ab   \n",
       "4  08acefb1-271e-01d2-1bb6-aae9f65dbb42   \n",
       "5  0b73e9c9-18c0-5c0e-c5d5-95d7ca56c050   \n",
       "6  0bb4b1f7-4128-89ef-a023-b81379c30f4f   \n",
       "7  0c8ad7c0-b403-98d1-d6c9-ff1dbbce7d7c   \n",
       "8  0f1c1042-f103-e85b-1a7a-03442861947d   \n",
       "9  14e884a7-dc93-7909-9585-2857c6eb29e2   \n",
       "\n",
       "                                           gene_info  \\\n",
       "0  'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...   \n",
       "1  'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...   \n",
       "2  'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...   \n",
       "3  'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...   \n",
       "4  'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...   \n",
       "5  'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...   \n",
       "6  'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...   \n",
       "7  'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...   \n",
       "8  'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...   \n",
       "9  'SOD3', None, 'TNF', 'LPL', 'F5', 'PON1', 'PPA...   \n",
       "\n",
       "                               clinical_significance  \\\n",
       "0  'Conflicting_interpretations_of_pathogenicity'...   \n",
       "1  'Conflicting_interpretations_of_pathogenicity'...   \n",
       "2  'Conflicting_interpretations_of_pathogenicity'...   \n",
       "3  'Conflicting_interpretations_of_pathogenicity'...   \n",
       "4  'Conflicting_interpretations_of_pathogenicity'...   \n",
       "5  'Conflicting_interpretations_of_pathogenicity'...   \n",
       "6  'Conflicting_interpretations_of_pathogenicity'...   \n",
       "7  'Conflicting_interpretations_of_pathogenicity'...   \n",
       "8  'Conflicting_interpretations_of_pathogenicity'...   \n",
       "9  'Conflicting_interpretations_of_pathogenicity'...   \n",
       "\n",
       "                                          contigname  \\\n",
       "0  '14', '9', '6', '16', '17', '8', '10', '11', '...   \n",
       "1  '14', '9', '6', '16', '17', '8', '10', '11', '...   \n",
       "2  '14', '9', '6', '16', '17', '8', '10', '11', '...   \n",
       "3  '14', '9', '6', '16', '17', '8', '10', '11', '...   \n",
       "4  '14', '9', '6', '16', '17', '8', '10', '11', '...   \n",
       "5  '14', '9', '6', '16', '17', '8', '10', '11', '...   \n",
       "6  '14', '9', '6', '16', '17', '8', '10', '11', '...   \n",
       "7  '14', '9', '6', '16', '17', '8', '10', '11', '...   \n",
       "8  '14', '9', '6', '16', '17', '8', '10', '11', '...   \n",
       "9  '14', '9', '6', '16', '17', '8', '10', '11', '...   \n",
       "\n",
       "                                               start     referenceallele  \\\n",
       "0  53767041, 44919688, 31667848, 12351625, 449086...  'T', 'G', 'A', 'C'   \n",
       "1  53767041, 44919688, 12351625, 31667848, 449086...  'T', 'G', 'A', 'C'   \n",
       "2  53767041, 44919688, 31667848, 12351625, 449086...  'T', 'G', 'A', 'C'   \n",
       "3  53767041, 44919688, 12351625, 31667848, 449086...  'T', 'G', 'A', 'C'   \n",
       "4  53767041, 44919688, 12351625, 31667848, 449086...  'T', 'G', 'A', 'C'   \n",
       "5  53767041, 44919688, 31667848, 12351625, 449086...  'T', 'G', 'A', 'C'   \n",
       "6  53767041, 44919688, 31667848, 12351625, 449086...  'T', 'G', 'A', 'C'   \n",
       "7  53767041, 44919688, 31667848, 12351625, 449086...  'T', 'G', 'A', 'C'   \n",
       "8  53767041, 44919688, 12351625, 31667848, 449086...  'T', 'G', 'A', 'C'   \n",
       "9  53767041, 44919688, 12351625, 31667848, 449086...  'T', 'G', 'A', 'C'   \n",
       "\n",
       "             alternatealleles phased               calls  \\\n",
       "0  '[A]', '[C]', '[G]', '[T]'  False  '[1, 1]', '[0, 1]'   \n",
       "1  '[A]', '[C]', '[G]', '[T]'  False  '[1, 1]', '[0, 1]'   \n",
       "2  '[A]', '[C]', '[G]', '[T]'  False  '[1, 1]', '[0, 1]'   \n",
       "3  '[A]', '[C]', '[G]', '[T]'  False  '[1, 1]', '[0, 1]'   \n",
       "4  '[A]', '[C]', '[G]', '[T]'  False  '[1, 1]', '[0, 1]'   \n",
       "5  '[A]', '[C]', '[G]', '[T]'  False  '[1, 1]', '[0, 1]'   \n",
       "6  '[A]', '[C]', '[G]', '[T]'  False  '[1, 1]', '[0, 1]'   \n",
       "7  '[A]', '[C]', '[G]', '[T]'  False  '[1, 1]', '[0, 1]'   \n",
       "8  '[A]', '[C]', '[G]', '[T]'  False  '[1, 1]', '[0, 1]'   \n",
       "9  '[A]', '[C]', '[G]', '[T]'  False  '[1, 1]', '[0, 1]'   \n",
       "\n",
       "  alzheimers_prediction  ... original_ngtdm_strength  \\\n",
       "0                     0  ...       8.547276310306838   \n",
       "1                     0  ...       9.241749116927542   \n",
       "2                     1  ...        9.02168678253902   \n",
       "3                     0  ...       8.776968231552623   \n",
       "4                     0  ...       8.601098896888011   \n",
       "5                     1  ...       9.106481386718237   \n",
       "6                     0  ...        9.38894545484775   \n",
       "7                     0  ...       9.202929567908852   \n",
       "8                     0  ...      11.269159030234578   \n",
       "9                     0  ...       8.875513496551257   \n",
       "\n",
       "                           imagesetid alzheimers_prediction.2  \\\n",
       "0  'd7298fe7dde4537b8343b5a702979aea'                       0   \n",
       "1  'e5e2ccf0487eed5523395431675fa708'                       0   \n",
       "2  '1a3c6b0044e2b58a13c106896da369ef'                       1   \n",
       "3  '41b14124cdd7f0ea3d459bcfe6120e7f'                       0   \n",
       "4  'dd9a6771bf229a91f03f608c336141de'                       0   \n",
       "5  '1526fea901654ad5e45f7d5f22377f81'                       1   \n",
       "6  'b94c9e7c782da3660ea6a905e2a91b34'                       0   \n",
       "7  'bccc8f07e6b97098dbb92cbafef2949c'                       0   \n",
       "8  'df83b4160e70a45bee0af99f9785d1eb'                       0   \n",
       "9  '20e77ca11e2b601b8ff9b9743efc1014'                       0   \n",
       "\n",
       "  coronary_heart_disease_prediction.2 stroke_prediction.2  \\\n",
       "0                                   0                   0   \n",
       "1                                   0                   0   \n",
       "2                                   0                   1   \n",
       "3                                   0                   0   \n",
       "4                                   0                   1   \n",
       "5                                   0                   1   \n",
       "6                                   0                   0   \n",
       "7                                   0                   1   \n",
       "8                                   0                   0   \n",
       "9                                   0                   1   \n",
       "\n",
       "  hypertension_prediction.2   eventtime.2             write_time.2  \\\n",
       "0                         0  1688074260.0  2023-06-29 21:37:00.939   \n",
       "1                         0  1688074260.0  2023-06-29 21:36:51.705   \n",
       "2                         1  1688074260.0  2023-06-29 21:37:00.886   \n",
       "3                         0  1688074260.0  2023-06-29 21:36:51.674   \n",
       "4                         0  1688074260.0  2023-06-29 21:36:59.693   \n",
       "5                         1  1688074260.0  2023-06-29 21:36:51.701   \n",
       "6                         0  1688074260.0  2023-06-29 21:37:00.006   \n",
       "7                         1  1688074260.0  2023-06-29 21:36:59.699   \n",
       "8                         0  1688074260.0  2023-06-29 21:36:59.693   \n",
       "9                         1  1688074260.0  2023-06-29 21:37:00.854   \n",
       "\n",
       "     api_invocation_time.2 is_deleted.2  \n",
       "0  2023-06-29 21:31:33.000        False  \n",
       "1  2023-06-29 21:31:33.000        False  \n",
       "2  2023-06-29 21:31:33.000        False  \n",
       "3  2023-06-29 21:31:33.000        False  \n",
       "4  2023-06-29 21:31:33.000        False  \n",
       "5  2023-06-29 21:31:33.000        False  \n",
       "6  2023-06-29 21:31:33.000        False  \n",
       "7  2023-06-29 21:31:33.000        False  \n",
       "8  2023-06-29 21:31:33.000        False  \n",
       "9  2023-06-29 21:31:33.000        False  \n",
       "\n",
       "[10 rows x 171 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "79588825-d49c-4309-b27e-7c79bc3f866c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ag' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-15-ffe180ae82e2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mag\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcore\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mutils\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrandom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mseed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m25\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'ag' is not defined"
     ]
    }
   ],
   "source": [
    "ag.core.utils.random.seed(25)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d2deae2-386a-49a0-9662-e81a1c17d550",
   "metadata": {},
   "source": [
    "## Alzheimers Prediction\n",
    "Splitting data for training and testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "544d6b79-7ea9-4082-bc95-cacd57015a13",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training size =  121\n",
      "Out of sample testing size =  30\n"
     ]
    }
   ],
   "source": [
    "#Alzheimers Prediction\n",
    "#Splitting data into training and testing 80:20\n",
    "dataset = dataset.loc[:, ~dataset.columns.str.startswith('diagnostics')]\n",
    "dataset = dataset.drop(columns = ['eventtime', 'write_time', 'api_invocation_time', 'is_deleted', 'eventtime.1', 'write_time.1', 'api_invocation_time.1', 'is_deleted.1', 'alzheimers_prediction.1',\n",
    "                                    'coronary_heart_disease_prediction.1', 'stroke_prediction.1', 'hypertension_prediction.1', 'patientid.1', 'eventtime.2', 'write_time.2', 'api_invocation_time.2', 'is_deleted.2', \n",
    "                                   'alzheimers_prediction.2', 'coronary_heart_disease_prediction.2', 'stroke_prediction.2', 'hypertension_prediction.2', 'patientid.2'])\n",
    "training= dataset.sample(frac=0.8, random_state=21)\n",
    "training = training.drop(columns = ['patientid', 'coronary_heart_disease_prediction', 'stroke_prediction', 'hypertension_prediction'])\n",
    "testing = dataset.drop(training.index)\n",
    "testing = testing.drop(columns = ['patientid', 'coronary_heart_disease_prediction', 'stroke_prediction', 'hypertension_prediction'])\n",
    "X_test = testing.drop(columns = ['alzheimers_prediction'])\n",
    "print(\"Training size = \", len(training))\n",
    "print(\"Out of sample testing size = \", len(testing))\n",
    "\n",
    "# Splitting data into training and testing for deploying to an endpoint\n",
    "training_target = training.pop(\"alzheimers_prediction\")\n",
    "training.insert(0, 'alzheimers_prediction', training_target)\n",
    "testing_target = testing.pop(\"alzheimers_prediction\")\n",
    "testing.insert(0, 'alzheimers_prediction', testing_target)\n",
    "training.to_csv('s3://multimodal-dataset-clinical-genomic-imaging/multimodal_alzheimers_training.csv', index=False)\n",
    "testing.to_csv('s3://multimodal-dataset-clinical-genomic-imaging/multimodal_alzheimers_testing.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0deacd5b-1dd8-4ee3-9b67-395e2298b52c",
   "metadata": {},
   "source": [
    "### Alzheimers prediction on clinical, genomic, and imaging data using Autogluon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "c9212862-66b2-4247-af4e-6711895c35f0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: path already exists! This predictor may overwrite an existing predictor! path=\"s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction\"\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/learner.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/predictor.pkl\n",
      "Beginning AutoGluon training ...\n",
      "AutoGluon will save models to \"s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/\"\n",
      "AutoGluon Version:  0.6.2\n",
      "Python Version:     3.7.10\n",
      "Operating System:   Linux\n",
      "Platform Machine:   x86_64\n",
      "Platform Version:   #1 SMP Thu Mar 23 09:54:12 UTC 2023\n",
      "Train Data Rows:    121\n",
      "Train Data Columns: 144\n",
      "Label Column: alzheimers_prediction\n",
      "Preprocessing data ...\n",
      "Selected class <--> label mapping:  class 1 = 1, class 0 = 0\n",
      "Using Feature Generators to preprocess the data ...\n",
      "Fitting AutoMLPipelineFeatureGenerator...\n",
      "\tAvailable Memory:                    2780.96 MB\n",
      "\tTrain Data (Original)  Memory Usage: 1.42 MB (0.1% of available memory)\n",
      "\tInferring data type of each feature based on column values. Set feature_metadata_in to manually specify special dtypes of the features.\n",
      "\tStage 1 Generators:\n",
      "\t\tFitting AsTypeFeatureGenerator...\n",
      "\t\t\tNote: Converting 8 features to boolean dtype as they only contain 2 unique values.\n",
      "\tStage 2 Generators:\n",
      "\t\tFitting FillNaFeatureGenerator...\n",
      "\tStage 3 Generators:\n",
      "\t\tFitting IdentityFeatureGenerator...\n",
      "\t\tFitting CategoryFeatureGenerator...\n",
      "\t\t\tFitting CategoryMemoryMinimizeFeatureGenerator...\n",
      "\t\tFitting DatetimeFeatureGenerator...\n",
      "/opt/conda/lib/python3.7/site-packages/autogluon/features/generators/datetime.py:59: FutureWarning: casting datetime64[ns, UTC] values to int64 with .astype(...) is deprecated and will raise in a future version. Use .view(...) instead.\n",
      "  good_rows = series[~series.isin(bad_rows)].astype(np.int64)\n",
      "\t\tFitting TextSpecialFeatureGenerator...\n",
      "\t\t\tFitting BinnedFeatureGenerator...\n",
      "\t\t\tFitting DropDuplicatesFeatureGenerator...\n",
      "\t\tFitting TextNgramFeatureGenerator...\n",
      "\t\t\tFitting CountVectorizer for text features: ['diagnostictype', 'diagnosticresult', 'conditionid', 'vaccinations', 'imagingname', 'imagebodypart', 'imagetitle', 'encounterreason', 'observationtype', 'observationtext', 'observationdisplay', 'proceduretype']\n",
      "\t\t\tCountVectorizer fit with vocabulary size = 1163\n",
      "\t\tWarning: Due to memory constraints, ngram feature count is being reduced. Allocate more memory to maximize model quality.\n",
      "\t\tReducing Vectorizer vocab size from 1163 to 728 to avoid OOM error\n",
      "\tStage 4 Generators:\n",
      "\t\tFitting DropUniqueFeatureGenerator...\n",
      "\tUseless Original Features (Count: 27): ['clinical_significance', 'contigname', 'referenceallele', 'alternatealleles', 'phased', 'deceasedindicator', 'encounterdiagnosis', 'hospitalizationstatus', 'medicationstatus', 'claim', 'original_shape_elongation', 'original_shape_flatness', 'original_shape_leastaxislength', 'original_shape_majoraxislength', 'original_shape_maximum2ddiametercolumn', 'original_shape_maximum2ddiameterrow', 'original_shape_maximum2ddiameterslice', 'original_shape_maximum3ddiameter', 'original_shape_meshvolume', 'original_shape_minoraxislength', 'original_shape_sphericity', 'original_shape_surfacearea', 'original_shape_surfacevolumeratio', 'original_shape_voxelvolume', 'original_firstorder_maximum', 'original_firstorder_minimum', 'original_firstorder_range']\n",
      "\t\tThese features carry no predictive signal and should be manually investigated.\n",
      "\t\tThis is typically a feature which has the same value for all rows.\n",
      "\t\tThese features do not need to be present at inference time.\n",
      "\tUnused Original Features (Count: 1): ['imagesetid']\n",
      "\t\tThese features were not used to generate any of the output features. Add a feature generator compatible with these features to utilize them.\n",
      "\t\tFeatures can also be unused if they carry very little information, such as being categorical but having almost entirely unique values or being duplicates of other features.\n",
      "\t\tThese features do not need to be present at inference time.\n",
      "\t\t('object', []) : 1 | ['imagesetid']\n",
      "\tTypes of features in original data (raw dtype, special dtypes):\n",
      "\t\t('float', [])                      : 91 | ['postalcode', 'original_firstorder_10percentile', 'original_firstorder_90percentile', 'original_firstorder_energy', 'original_firstorder_entropy', ...]\n",
      "\t\t('object', [])                     :  7 | ['gender', 'maritalstatus', 'deceaseddatetime', 'categorycode', 'categorydescription', ...]\n",
      "\t\t('object', ['datetime_as_object']) :  1 | ['medicationdate']\n",
      "\t\t('object', ['text'])               : 17 | ['gene_info', 'start', 'calls', 'diagnostictype', 'diagnosticcategory', ...]\n",
      "\tTypes of features in processed data (raw dtype, special dtypes):\n",
      "\t\t('category', [])                    :   4 | ['gender', 'maritalstatus', 'deceaseddatetime', 'imagecode']\n",
      "\t\t('category', ['text_as_category'])  :  11 | ['diagnostictype', 'diagnosticresult', 'conditionid', 'vaccinations', 'imagingname', ...]\n",
      "\t\t('float', [])                       :  91 | ['postalcode', 'original_firstorder_10percentile', 'original_firstorder_90percentile', 'original_firstorder_energy', 'original_firstorder_entropy', ...]\n",
      "\t\t('int', ['binned', 'text_special']) :  92 | ['diagnostictype.char_count', 'diagnostictype.word_count', 'diagnostictype.capital_ratio', 'diagnostictype.lower_ratio', 'diagnostictype.digit_ratio', ...]\n",
      "\t\t('int', ['bool'])                   :   8 | ['gene_info', 'start', 'calls', 'diagnosticcategory', 'categorycode', ...]\n",
      "\t\t('int', ['datetime_as_int'])        :   5 | ['medicationdate', 'medicationdate.year', 'medicationdate.month', 'medicationdate.day', 'medicationdate.dayofweek']\n",
      "\t\t('int', ['text_ngram'])             : 719 | ['__nlp__.10', '__nlp__.10 verbal', '__nlp__.10 verbal numeric', '__nlp__.2000', '__nlp__.2000 panel', ...]\n",
      "\t3.3s = Fit runtime\n",
      "\t116 features in original data used to generate 930 features in processed data.\n",
      "\tTrain Data (Processed) Memory Usage: 0.29 MB (0.0% of available memory)\n",
      "Data preprocessing and feature engineering runtime = 3.42s ...\n",
      "AutoGluon will gauge predictive performance using evaluation metric: 'accuracy'\n",
      "\tTo change this, specify the eval_metric parameter of Predictor()\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/learner.pkl\n",
      "Automatically generating train/validation split with holdout_frac=0.1, Train Rows: 108, Val Rows: 13\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/utils/data/X.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/utils/data/y.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/utils/data/X_val.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/utils/data/y_val.pkl\n",
      "Excluded Model Types: ['CAT', 'XGB']\n",
      "\tFound 'CAT' model in hyperparameters, but 'CAT' is present in `excluded_model_types` and will be removed.\n",
      "\tFound 'XGB' model in hyperparameters, but 'XGB' is present in `excluded_model_types` and will be removed.\n",
      "Fitting 11 L1 models ...\n",
      "Fitting model: KNeighborsUnif ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/KNeighborsUnif/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/utils/attr/KNeighborsUnif/y_pred_proba_val.pkl\n",
      "\t0.9231\t = Validation score   (accuracy)\n",
      "\t0.1s\t = Training   runtime\n",
      "\t0.1s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/trainer.pkl\n",
      "Fitting model: KNeighborsDist ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/KNeighborsDist/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/utils/attr/KNeighborsDist/y_pred_proba_val.pkl\n",
      "\t0.9231\t = Validation score   (accuracy)\n",
      "\t0.09s\t = Training   runtime\n",
      "\t0.1s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/trainer.pkl\n",
      "Fitting model: LightGBMXT ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/LightGBMXT/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/utils/attr/LightGBMXT/y_pred_proba_val.pkl\n",
      "\t1.0\t = Validation score   (accuracy)\n",
      "\t0.93s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/trainer.pkl\n",
      "Fitting model: LightGBM ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/LightGBM/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/utils/attr/LightGBM/y_pred_proba_val.pkl\n",
      "\t0.9231\t = Validation score   (accuracy)\n",
      "\t0.63s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/trainer.pkl\n",
      "Fitting model: RandomForestGini ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/RandomForestGini/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/utils/attr/RandomForestGini/y_pred_proba_val.pkl\n",
      "\t0.9231\t = Validation score   (accuracy)\n",
      "\t0.88s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/trainer.pkl\n",
      "Fitting model: RandomForestEntr ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/RandomForestEntr/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/utils/attr/RandomForestEntr/y_pred_proba_val.pkl\n",
      "\t0.9231\t = Validation score   (accuracy)\n",
      "\t0.88s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/trainer.pkl\n",
      "Fitting model: ExtraTreesGini ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/ExtraTreesGini/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/utils/attr/ExtraTreesGini/y_pred_proba_val.pkl\n",
      "\t0.9231\t = Validation score   (accuracy)\n",
      "\t0.89s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/trainer.pkl\n",
      "Fitting model: ExtraTreesEntr ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/ExtraTreesEntr/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/utils/attr/ExtraTreesEntr/y_pred_proba_val.pkl\n",
      "\t0.9231\t = Validation score   (accuracy)\n",
      "\t0.91s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/trainer.pkl\n",
      "Fitting model: NeuralNetFastAI ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/NeuralNetFastAI/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/NeuralNetFastAI/model-internals.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/utils/attr/NeuralNetFastAI/y_pred_proba_val.pkl\n",
      "\t0.9231\t = Validation score   (accuracy)\n",
      "\t1.32s\t = Training   runtime\n",
      "\t0.06s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/trainer.pkl\n",
      "Fitting model: NeuralNetTorch ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/NeuralNetTorch/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/utils/attr/NeuralNetTorch/y_pred_proba_val.pkl\n",
      "\t0.9231\t = Validation score   (accuracy)\n",
      "\t1.33s\t = Training   runtime\n",
      "\t0.04s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/trainer.pkl\n",
      "Fitting model: LightGBMLarge ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/LightGBMLarge/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/utils/attr/LightGBMLarge/y_pred_proba_val.pkl\n",
      "\t0.9231\t = Validation score   (accuracy)\n",
      "\t1.58s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/trainer.pkl\n",
      "Fitting model: WeightedEnsemble_L2 ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/WeightedEnsemble_L2/utils/model_template.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/WeightedEnsemble_L2/utils/oof.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/WeightedEnsemble_L2/model.pkl\n",
      "\t1.0\t = Validation score   (accuracy)\n",
      "\t0.25s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/trainer.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/trainer.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/trainer.pkl\n",
      "AutoGluon training complete, total runtime = 20.18s ... Best model: \"WeightedEnsemble_L2\"\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/models/trainer.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/learner.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/predictor.pkl\n",
      "TabularPredictor saved. To load, use: predictor = TabularPredictor.load(\"s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-alzheimers-prediction/\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Training time= 21.408601760864258 seconds ---\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "start_time = time.time()\n",
    "buckt = sm_session.default_bucket()\n",
    "prefix= \"genomic-clinical-imaging-alzheimers-prediction\"\n",
    "save_file = 's3://{}/{}'.format(buckt, prefix)\n",
    "predictor = TabularPredictor(label= 'alzheimers_prediction', problem_type= 'binary', path=save_file).fit(train_data=training, holdout_frac=0.1, excluded_model_types=['CAT', 'XGB'])\n",
    "print(\"--- Training time= %s seconds ---\" % (time.time() - start_time))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "3a8cc026-7387-4b82-9a9c-1bc2b7818a94",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Evaluation: accuracy on test data: 0.8\n",
      "Evaluations on test data:\n",
      "{\n",
      "    \"accuracy\": 0.8,\n",
      "    \"balanced_accuracy\": 0.625,\n",
      "    \"mcc\": 0.44320263021395917,\n",
      "    \"f1\": 0.4,\n",
      "    \"precision\": 1.0,\n",
      "    \"recall\": 0.25\n",
      "}\n",
      "Detailed (per-class) classification report:\n",
      "{\n",
      "    \"0\": {\n",
      "        \"precision\": 0.7857142857142857,\n",
      "        \"recall\": 1.0,\n",
      "        \"f1-score\": 0.88,\n",
      "        \"support\": 22\n",
      "    },\n",
      "    \"1\": {\n",
      "        \"precision\": 1.0,\n",
      "        \"recall\": 0.25,\n",
      "        \"f1-score\": 0.4,\n",
      "        \"support\": 8\n",
      "    },\n",
      "    \"accuracy\": 0.8,\n",
      "    \"macro avg\": {\n",
      "        \"precision\": 0.8928571428571428,\n",
      "        \"recall\": 0.625,\n",
      "        \"f1-score\": 0.64,\n",
      "        \"support\": 30\n",
      "    },\n",
      "    \"weighted avg\": {\n",
      "        \"precision\": 0.8428571428571429,\n",
      "        \"recall\": 0.8,\n",
      "        \"f1-score\": 0.752,\n",
      "        \"support\": 30\n",
      "    }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'accuracy': 0.8,\n",
       " 'balanced_accuracy': 0.625,\n",
       " 'mcc': 0.44320263021395917,\n",
       " 'f1': 0.4,\n",
       " 'precision': 1.0,\n",
       " 'recall': 0.25,\n",
       " 'confusion_matrix':     0  1\n",
       " 0  22  0\n",
       " 1   6  2,\n",
       " 'classification_report': {'0': {'precision': 0.7857142857142857,\n",
       "   'recall': 1.0,\n",
       "   'f1-score': 0.88,\n",
       "   'support': 22},\n",
       "  '1': {'precision': 1.0, 'recall': 0.25, 'f1-score': 0.4, 'support': 8},\n",
       "  'accuracy': 0.8,\n",
       "  'macro avg': {'precision': 0.8928571428571428,\n",
       "   'recall': 0.625,\n",
       "   'f1-score': 0.64,\n",
       "   'support': 30},\n",
       "  'weighted avg': {'precision': 0.8428571428571429,\n",
       "   'recall': 0.8,\n",
       "   'f1-score': 0.752,\n",
       "   'support': 30}}}"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictor.evaluate_predictions(y_true=testing['alzheimers_prediction'], y_pred=predictor.predict(X_test), auxiliary_metrics=True, detailed_report=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1630bd6-ff2f-4b36-be9b-c8b30b6bbf70",
   "metadata": {},
   "source": [
    "## Coronary heart disease Prediction\n",
    "Splitting data for training and testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "accbdc2f-b152-46fb-ab09-f6ce919a8c95",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training size =  121\n",
      "Out of sample testing size =  30\n"
     ]
    }
   ],
   "source": [
    "#coronary_heart_disease_prediction\n",
    "#Splitting data into training and testing 80:20\n",
    "training = dataset.sample(frac=0.8, random_state=25)\n",
    "training =  training.drop(columns = ['patientid', 'alzheimers_prediction', 'stroke_prediction', 'hypertension_prediction'])\n",
    "testing = dataset.drop(training.index)\n",
    "testing = testing.drop(columns = ['patientid', 'alzheimers_prediction', 'stroke_prediction', 'hypertension_prediction'])\n",
    "X_test = testing.drop(columns = ['coronary_heart_disease_prediction'])\n",
    "print(\"Training size = \", len(training))\n",
    "print(\"Out of sample testing size = \", len(testing))\n",
    "\n",
    "# Splitting data into training and testing for deploying to an endpoint\n",
    "training_target = training.pop(\"coronary_heart_disease_prediction\")\n",
    "training.insert(0, 'coronary_heart_disease_prediction', training_target)\n",
    "testing_target = testing.pop(\"coronary_heart_disease_prediction\")\n",
    "testing.insert(0, 'coronary_heart_disease_prediction', testing_target)\n",
    "training.to_csv('s3://multimodal-dataset-clinical-genomic-imaging/multimodal_coronary_heart_disease_training.csv', index=False)\n",
    "testing.to_csv('s3://multimodal-dataset-clinical-genomic-imaging/multimodal_coronary_heart_disease_testing.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "f2a7e896-470a-4842-855c-e4e949102ea0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Coronary heart disease prediction on clinical,  genomic, and imaging data using Autogluon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "f5e09744-d07b-4a89-93a8-aeaf2e6d7e3d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: path already exists! This predictor may overwrite an existing predictor! path=\"s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction\"\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/learner.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/predictor.pkl\n",
      "Beginning AutoGluon training ...\n",
      "AutoGluon will save models to \"s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/\"\n",
      "AutoGluon Version:  0.6.2\n",
      "Python Version:     3.7.10\n",
      "Operating System:   Linux\n",
      "Platform Machine:   x86_64\n",
      "Platform Version:   #1 SMP Thu Mar 23 09:54:12 UTC 2023\n",
      "Train Data Rows:    121\n",
      "Train Data Columns: 144\n",
      "Label Column: coronary_heart_disease_prediction\n",
      "Preprocessing data ...\n",
      "Selected class <--> label mapping:  class 1 = 1, class 0 = 0\n",
      "Using Feature Generators to preprocess the data ...\n",
      "Fitting AutoMLPipelineFeatureGenerator...\n",
      "\tAvailable Memory:                    2779.88 MB\n",
      "\tTrain Data (Original)  Memory Usage: 1.52 MB (0.1% of available memory)\n",
      "\tInferring data type of each feature based on column values. Set feature_metadata_in to manually specify special dtypes of the features.\n",
      "\tStage 1 Generators:\n",
      "\t\tFitting AsTypeFeatureGenerator...\n",
      "\t\t\tNote: Converting 8 features to boolean dtype as they only contain 2 unique values.\n",
      "\tStage 2 Generators:\n",
      "\t\tFitting FillNaFeatureGenerator...\n",
      "\tStage 3 Generators:\n",
      "\t\tFitting IdentityFeatureGenerator...\n",
      "\t\tFitting CategoryFeatureGenerator...\n",
      "\t\t\tFitting CategoryMemoryMinimizeFeatureGenerator...\n",
      "\t\tFitting DatetimeFeatureGenerator...\n",
      "/opt/conda/lib/python3.7/site-packages/autogluon/features/generators/datetime.py:59: FutureWarning: casting datetime64[ns, UTC] values to int64 with .astype(...) is deprecated and will raise in a future version. Use .view(...) instead.\n",
      "  good_rows = series[~series.isin(bad_rows)].astype(np.int64)\n",
      "\t\tFitting TextSpecialFeatureGenerator...\n",
      "\t\t\tFitting BinnedFeatureGenerator...\n",
      "\t\t\tFitting DropDuplicatesFeatureGenerator...\n",
      "\t\tFitting TextNgramFeatureGenerator...\n",
      "\t\t\tFitting CountVectorizer for text features: ['diagnostictype', 'diagnosticresult', 'conditionid', 'vaccinations', 'imagingname', 'imagebodypart', 'imagetitle', 'encounterreason', 'observationtype', 'observationtext', 'observationdisplay', 'proceduretype']\n",
      "\t\t\tCountVectorizer fit with vocabulary size = 1161\n",
      "\t\tWarning: Due to memory constraints, ngram feature count is being reduced. Allocate more memory to maximize model quality.\n",
      "\t\tReducing Vectorizer vocab size from 1161 to 726 to avoid OOM error\n",
      "\tStage 4 Generators:\n",
      "\t\tFitting DropUniqueFeatureGenerator...\n",
      "\tUseless Original Features (Count: 27): ['clinical_significance', 'contigname', 'referenceallele', 'alternatealleles', 'phased', 'deceasedindicator', 'encounterdiagnosis', 'hospitalizationstatus', 'medicationstatus', 'claim', 'original_shape_elongation', 'original_shape_flatness', 'original_shape_leastaxislength', 'original_shape_majoraxislength', 'original_shape_maximum2ddiametercolumn', 'original_shape_maximum2ddiameterrow', 'original_shape_maximum2ddiameterslice', 'original_shape_maximum3ddiameter', 'original_shape_meshvolume', 'original_shape_minoraxislength', 'original_shape_sphericity', 'original_shape_surfacearea', 'original_shape_surfacevolumeratio', 'original_shape_voxelvolume', 'original_firstorder_maximum', 'original_firstorder_minimum', 'original_firstorder_range']\n",
      "\t\tThese features carry no predictive signal and should be manually investigated.\n",
      "\t\tThis is typically a feature which has the same value for all rows.\n",
      "\t\tThese features do not need to be present at inference time.\n",
      "\tUnused Original Features (Count: 1): ['imagesetid']\n",
      "\t\tThese features were not used to generate any of the output features. Add a feature generator compatible with these features to utilize them.\n",
      "\t\tFeatures can also be unused if they carry very little information, such as being categorical but having almost entirely unique values or being duplicates of other features.\n",
      "\t\tThese features do not need to be present at inference time.\n",
      "\t\t('object', []) : 1 | ['imagesetid']\n",
      "\tTypes of features in original data (raw dtype, special dtypes):\n",
      "\t\t('float', [])                      : 91 | ['postalcode', 'original_firstorder_10percentile', 'original_firstorder_90percentile', 'original_firstorder_energy', 'original_firstorder_entropy', ...]\n",
      "\t\t('object', [])                     :  6 | ['gender', 'maritalstatus', 'categorycode', 'categorydescription', 'imagecode', ...]\n",
      "\t\t('object', ['datetime_as_object']) :  2 | ['deceaseddatetime', 'medicationdate']\n",
      "\t\t('object', ['text'])               : 17 | ['gene_info', 'start', 'calls', 'diagnostictype', 'diagnosticcategory', ...]\n",
      "\tTypes of features in processed data (raw dtype, special dtypes):\n",
      "\t\t('category', [])                    :   3 | ['gender', 'maritalstatus', 'imagecode']\n",
      "\t\t('category', ['text_as_category'])  :  11 | ['diagnostictype', 'diagnosticresult', 'conditionid', 'vaccinations', 'imagingname', ...]\n",
      "\t\t('float', [])                       :  91 | ['postalcode', 'original_firstorder_10percentile', 'original_firstorder_90percentile', 'original_firstorder_energy', 'original_firstorder_entropy', ...]\n",
      "\t\t('int', ['binned', 'text_special']) :  92 | ['diagnostictype.char_count', 'diagnostictype.word_count', 'diagnostictype.capital_ratio', 'diagnostictype.lower_ratio', 'diagnostictype.digit_ratio', ...]\n",
      "\t\t('int', ['bool'])                   :   8 | ['gene_info', 'start', 'calls', 'diagnosticcategory', 'categorycode', ...]\n",
      "\t\t('int', ['datetime_as_int'])        :  10 | ['deceaseddatetime', 'deceaseddatetime.year', 'deceaseddatetime.month', 'deceaseddatetime.day', 'deceaseddatetime.dayofweek', ...]\n",
      "\t\t('int', ['text_ngram'])             : 717 | ['__nlp__.10', '__nlp__.10 verbal', '__nlp__.10 verbal numeric', '__nlp__.2000', '__nlp__.2000 panel', ...]\n",
      "\t3.2s = Fit runtime\n",
      "\t116 features in original data used to generate 932 features in processed data.\n",
      "\tTrain Data (Processed) Memory Usage: 0.29 MB (0.0% of available memory)\n",
      "Data preprocessing and feature engineering runtime = 3.28s ...\n",
      "AutoGluon will gauge predictive performance using evaluation metric: 'accuracy'\n",
      "\tTo change this, specify the eval_metric parameter of Predictor()\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/learner.pkl\n",
      "Automatically generating train/validation split with holdout_frac=0.1, Train Rows: 108, Val Rows: 13\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/utils/data/X.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/utils/data/y.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/utils/data/X_val.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/utils/data/y_val.pkl\n",
      "Excluded Model Types: ['CAT', 'XGB']\n",
      "\tFound 'CAT' model in hyperparameters, but 'CAT' is present in `excluded_model_types` and will be removed.\n",
      "\tFound 'XGB' model in hyperparameters, but 'XGB' is present in `excluded_model_types` and will be removed.\n",
      "Fitting 11 L1 models ...\n",
      "Fitting model: KNeighborsUnif ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/KNeighborsUnif/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/utils/attr/KNeighborsUnif/y_pred_proba_val.pkl\n",
      "\t0.7692\t = Validation score   (accuracy)\n",
      "\t0.09s\t = Training   runtime\n",
      "\t0.1s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/trainer.pkl\n",
      "Fitting model: KNeighborsDist ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/KNeighborsDist/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/utils/attr/KNeighborsDist/y_pred_proba_val.pkl\n",
      "\t0.7692\t = Validation score   (accuracy)\n",
      "\t0.11s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/trainer.pkl\n",
      "Fitting model: LightGBMXT ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/LightGBMXT/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/utils/attr/LightGBMXT/y_pred_proba_val.pkl\n",
      "\t0.8462\t = Validation score   (accuracy)\n",
      "\t1.08s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/trainer.pkl\n",
      "Fitting model: LightGBM ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/LightGBM/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/utils/attr/LightGBM/y_pred_proba_val.pkl\n",
      "\t0.8462\t = Validation score   (accuracy)\n",
      "\t0.64s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/trainer.pkl\n",
      "Fitting model: RandomForestGini ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/RandomForestGini/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/utils/attr/RandomForestGini/y_pred_proba_val.pkl\n",
      "\t0.8462\t = Validation score   (accuracy)\n",
      "\t0.88s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/trainer.pkl\n",
      "Fitting model: RandomForestEntr ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/RandomForestEntr/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/utils/attr/RandomForestEntr/y_pred_proba_val.pkl\n",
      "\t0.7692\t = Validation score   (accuracy)\n",
      "\t0.89s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/trainer.pkl\n",
      "Fitting model: ExtraTreesGini ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/ExtraTreesGini/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/utils/attr/ExtraTreesGini/y_pred_proba_val.pkl\n",
      "\t0.8462\t = Validation score   (accuracy)\n",
      "\t0.92s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/trainer.pkl\n",
      "Fitting model: ExtraTreesEntr ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/ExtraTreesEntr/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/utils/attr/ExtraTreesEntr/y_pred_proba_val.pkl\n",
      "\t0.7692\t = Validation score   (accuracy)\n",
      "\t0.88s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/trainer.pkl\n",
      "Fitting model: NeuralNetFastAI ...\n",
      "No improvement since epoch 7: early stopping\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/NeuralNetFastAI/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/NeuralNetFastAI/model-internals.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/utils/attr/NeuralNetFastAI/y_pred_proba_val.pkl\n",
      "\t0.9231\t = Validation score   (accuracy)\n",
      "\t1.3s\t = Training   runtime\n",
      "\t0.06s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/trainer.pkl\n",
      "Fitting model: NeuralNetTorch ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/NeuralNetTorch/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/utils/attr/NeuralNetTorch/y_pred_proba_val.pkl\n",
      "\t0.8462\t = Validation score   (accuracy)\n",
      "\t1.33s\t = Training   runtime\n",
      "\t0.03s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/trainer.pkl\n",
      "Fitting model: LightGBMLarge ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/LightGBMLarge/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/utils/attr/LightGBMLarge/y_pred_proba_val.pkl\n",
      "\t0.8462\t = Validation score   (accuracy)\n",
      "\t1.44s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/trainer.pkl\n",
      "Fitting model: WeightedEnsemble_L2 ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/WeightedEnsemble_L2/utils/model_template.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/WeightedEnsemble_L2/utils/oof.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/WeightedEnsemble_L2/model.pkl\n",
      "\t0.9231\t = Validation score   (accuracy)\n",
      "\t0.28s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/trainer.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/trainer.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/trainer.pkl\n",
      "AutoGluon training complete, total runtime = 20.23s ... Best model: \"WeightedEnsemble_L2\"\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/models/trainer.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/learner.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/predictor.pkl\n",
      "TabularPredictor saved. To load, use: predictor = TabularPredictor.load(\"s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-coronary-heart-disease-prediction/\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Training time= 21.052737712860107 seconds ---\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "start_time = time.time()\n",
    "buckt = sm_session.default_bucket()\n",
    "prefix= \"genomic-clinical-imaging-coronary-heart-disease-prediction\"\n",
    "save_file = 's3://{}/{}'.format(buckt, prefix)\n",
    "predictor = TabularPredictor(label= 'coronary_heart_disease_prediction', problem_type= 'binary', path=save_file).fit(train_data=training, holdout_frac=0.1, excluded_model_types=['CAT', 'XGB'])\n",
    "print(\"--- Training time= %s seconds ---\" % (time.time() - start_time))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "9eacb091-7a47-4efc-a360-4a265efbfe0c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Evaluation: accuracy on test data: 0.8333333333333334\n",
      "Evaluations on test data:\n",
      "{\n",
      "    \"accuracy\": 0.8333333333333334,\n",
      "    \"balanced_accuracy\": 0.6923076923076923,\n",
      "    \"mcc\": 0.35082320772281167,\n",
      "    \"f1\": 0.4444444444444445,\n",
      "    \"precision\": 0.4,\n",
      "    \"recall\": 0.5\n",
      "}\n",
      "Detailed (per-class) classification report:\n",
      "{\n",
      "    \"0\": {\n",
      "        \"precision\": 0.92,\n",
      "        \"recall\": 0.8846153846153846,\n",
      "        \"f1-score\": 0.9019607843137256,\n",
      "        \"support\": 26\n",
      "    },\n",
      "    \"1\": {\n",
      "        \"precision\": 0.4,\n",
      "        \"recall\": 0.5,\n",
      "        \"f1-score\": 0.4444444444444445,\n",
      "        \"support\": 4\n",
      "    },\n",
      "    \"accuracy\": 0.8333333333333334,\n",
      "    \"macro avg\": {\n",
      "        \"precision\": 0.66,\n",
      "        \"recall\": 0.6923076923076923,\n",
      "        \"f1-score\": 0.673202614379085,\n",
      "        \"support\": 30\n",
      "    },\n",
      "    \"weighted avg\": {\n",
      "        \"precision\": 0.8506666666666668,\n",
      "        \"recall\": 0.8333333333333334,\n",
      "        \"f1-score\": 0.8409586056644881,\n",
      "        \"support\": 30\n",
      "    }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'accuracy': 0.8333333333333334,\n",
       " 'balanced_accuracy': 0.6923076923076923,\n",
       " 'mcc': 0.35082320772281167,\n",
       " 'f1': 0.4444444444444445,\n",
       " 'precision': 0.4,\n",
       " 'recall': 0.5,\n",
       " 'confusion_matrix':     0  1\n",
       " 0  23  3\n",
       " 1   2  2,\n",
       " 'classification_report': {'0': {'precision': 0.92,\n",
       "   'recall': 0.8846153846153846,\n",
       "   'f1-score': 0.9019607843137256,\n",
       "   'support': 26},\n",
       "  '1': {'precision': 0.4,\n",
       "   'recall': 0.5,\n",
       "   'f1-score': 0.4444444444444445,\n",
       "   'support': 4},\n",
       "  'accuracy': 0.8333333333333334,\n",
       "  'macro avg': {'precision': 0.66,\n",
       "   'recall': 0.6923076923076923,\n",
       "   'f1-score': 0.673202614379085,\n",
       "   'support': 30},\n",
       "  'weighted avg': {'precision': 0.8506666666666668,\n",
       "   'recall': 0.8333333333333334,\n",
       "   'f1-score': 0.8409586056644881,\n",
       "   'support': 30}}}"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictor.evaluate_predictions(y_true=testing['coronary_heart_disease_prediction'], y_pred=predictor.predict(X_test), auxiliary_metrics=True, detailed_report=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7deb337c-9b26-4d70-b700-c8eda40fa115",
   "metadata": {},
   "source": [
    "## Stroke Prediction\n",
    "Splitting data for training and testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b0ef9c2d-eafc-4f05-a155-eb5237eb5052",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training size =  121\n",
      "Out of sample testing size =  30\n"
     ]
    }
   ],
   "source": [
    "#stroke_prediction\n",
    "#Splitting data into training and testing 80:20\n",
    "training = dataset.sample(frac=0.8, random_state=30)\n",
    "training =  training.drop(columns = ['patientid', 'alzheimers_prediction', 'coronary_heart_disease_prediction', 'hypertension_prediction'])\n",
    "testing = dataset.drop(training.index)\n",
    "testing = testing.drop(columns = ['patientid', 'alzheimers_prediction', 'coronary_heart_disease_prediction', 'hypertension_prediction'])\n",
    "X_test = testing.drop(columns = ['stroke_prediction'])\n",
    "print(\"Training size = \", len(training))\n",
    "print(\"Out of sample testing size = \", len(testing))\n",
    "\n",
    "# Splitting data into training and testing for deploying to an endpoint\n",
    "training_target = training.pop(\"stroke_prediction\")\n",
    "training.insert(0, 'stroke_prediction', training_target)\n",
    "testing_target = testing.pop(\"stroke_prediction\")\n",
    "testing.insert(0, 'stroke_prediction', testing_target)\n",
    "training.to_csv('s3://multimodal-dataset-clinical-genomic-imaging/multimodal_stroke_training.csv', index=False)\n",
    "testing.to_csv('s3://multimodal-dataset-clinical-genomic-imaging/multimodal_stroke_testing.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a753dc2-452f-44d0-92af-2c335a2478a9",
   "metadata": {},
   "source": [
    "### Stroke prediction on clinical, genomic, and imaging data using Autogluon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "8bc9ec99-5f75-486e-982a-38d57fa4b97d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: path already exists! This predictor may overwrite an existing predictor! path=\"s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction\"\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/learner.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/predictor.pkl\n",
      "Beginning AutoGluon training ...\n",
      "AutoGluon will save models to \"s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/\"\n",
      "AutoGluon Version:  0.6.2\n",
      "Python Version:     3.7.10\n",
      "Operating System:   Linux\n",
      "Platform Machine:   x86_64\n",
      "Platform Version:   #1 SMP Thu Mar 23 09:54:12 UTC 2023\n",
      "Train Data Rows:    121\n",
      "Train Data Columns: 144\n",
      "Label Column: stroke_prediction\n",
      "Preprocessing data ...\n",
      "Selected class <--> label mapping:  class 1 = 1, class 0 = 0\n",
      "Using Feature Generators to preprocess the data ...\n",
      "Fitting AutoMLPipelineFeatureGenerator...\n",
      "\tAvailable Memory:                    2778.9 MB\n",
      "\tTrain Data (Original)  Memory Usage: 1.38 MB (0.0% of available memory)\n",
      "\tInferring data type of each feature based on column values. Set feature_metadata_in to manually specify special dtypes of the features.\n",
      "\tStage 1 Generators:\n",
      "\t\tFitting AsTypeFeatureGenerator...\n",
      "\t\t\tNote: Converting 8 features to boolean dtype as they only contain 2 unique values.\n",
      "\tStage 2 Generators:\n",
      "\t\tFitting FillNaFeatureGenerator...\n",
      "\tStage 3 Generators:\n",
      "\t\tFitting IdentityFeatureGenerator...\n",
      "\t\tFitting CategoryFeatureGenerator...\n",
      "\t\t\tFitting CategoryMemoryMinimizeFeatureGenerator...\n",
      "\t\tFitting DatetimeFeatureGenerator...\n",
      "/opt/conda/lib/python3.7/site-packages/autogluon/features/generators/datetime.py:59: FutureWarning: casting datetime64[ns, UTC] values to int64 with .astype(...) is deprecated and will raise in a future version. Use .view(...) instead.\n",
      "  good_rows = series[~series.isin(bad_rows)].astype(np.int64)\n",
      "\t\tFitting TextSpecialFeatureGenerator...\n",
      "\t\t\tFitting BinnedFeatureGenerator...\n",
      "\t\t\tFitting DropDuplicatesFeatureGenerator...\n",
      "\t\tFitting TextNgramFeatureGenerator...\n",
      "\t\t\tFitting CountVectorizer for text features: ['diagnostictype', 'diagnosticresult', 'conditionid', 'vaccinations', 'imagingname', 'imagebodypart', 'imagetitle', 'encounterreason', 'observationtype', 'observationtext', 'observationdisplay', 'proceduretype']\n",
      "\t\t\tCountVectorizer fit with vocabulary size = 1157\n",
      "\t\tWarning: Due to memory constraints, ngram feature count is being reduced. Allocate more memory to maximize model quality.\n",
      "\t\tReducing Vectorizer vocab size from 1157 to 722 to avoid OOM error\n",
      "\tStage 4 Generators:\n",
      "\t\tFitting DropUniqueFeatureGenerator...\n",
      "\tUseless Original Features (Count: 27): ['clinical_significance', 'contigname', 'referenceallele', 'alternatealleles', 'phased', 'deceasedindicator', 'encounterdiagnosis', 'hospitalizationstatus', 'medicationstatus', 'claim', 'original_shape_elongation', 'original_shape_flatness', 'original_shape_leastaxislength', 'original_shape_majoraxislength', 'original_shape_maximum2ddiametercolumn', 'original_shape_maximum2ddiameterrow', 'original_shape_maximum2ddiameterslice', 'original_shape_maximum3ddiameter', 'original_shape_meshvolume', 'original_shape_minoraxislength', 'original_shape_sphericity', 'original_shape_surfacearea', 'original_shape_surfacevolumeratio', 'original_shape_voxelvolume', 'original_firstorder_maximum', 'original_firstorder_minimum', 'original_firstorder_range']\n",
      "\t\tThese features carry no predictive signal and should be manually investigated.\n",
      "\t\tThis is typically a feature which has the same value for all rows.\n",
      "\t\tThese features do not need to be present at inference time.\n",
      "\tUnused Original Features (Count: 1): ['imagesetid']\n",
      "\t\tThese features were not used to generate any of the output features. Add a feature generator compatible with these features to utilize them.\n",
      "\t\tFeatures can also be unused if they carry very little information, such as being categorical but having almost entirely unique values or being duplicates of other features.\n",
      "\t\tThese features do not need to be present at inference time.\n",
      "\t\t('object', []) : 1 | ['imagesetid']\n",
      "\tTypes of features in original data (raw dtype, special dtypes):\n",
      "\t\t('float', [])                      : 91 | ['postalcode', 'original_firstorder_10percentile', 'original_firstorder_90percentile', 'original_firstorder_energy', 'original_firstorder_entropy', ...]\n",
      "\t\t('object', [])                     :  7 | ['gender', 'maritalstatus', 'deceaseddatetime', 'categorycode', 'categorydescription', ...]\n",
      "\t\t('object', ['datetime_as_object']) :  1 | ['medicationdate']\n",
      "\t\t('object', ['text'])               : 17 | ['gene_info', 'start', 'calls', 'diagnostictype', 'diagnosticcategory', ...]\n",
      "\tTypes of features in processed data (raw dtype, special dtypes):\n",
      "\t\t('category', [])                    :   4 | ['gender', 'maritalstatus', 'deceaseddatetime', 'imagecode']\n",
      "\t\t('category', ['text_as_category'])  :  11 | ['diagnostictype', 'diagnosticresult', 'conditionid', 'vaccinations', 'imagingname', ...]\n",
      "\t\t('float', [])                       :  91 | ['postalcode', 'original_firstorder_10percentile', 'original_firstorder_90percentile', 'original_firstorder_energy', 'original_firstorder_entropy', ...]\n",
      "\t\t('int', ['binned', 'text_special']) :  91 | ['diagnostictype.char_count', 'diagnostictype.word_count', 'diagnostictype.capital_ratio', 'diagnostictype.lower_ratio', 'diagnostictype.digit_ratio', ...]\n",
      "\t\t('int', ['bool'])                   :   8 | ['gene_info', 'start', 'calls', 'diagnosticcategory', 'categorycode', ...]\n",
      "\t\t('int', ['datetime_as_int'])        :   5 | ['medicationdate', 'medicationdate.year', 'medicationdate.month', 'medicationdate.day', 'medicationdate.dayofweek']\n",
      "\t\t('int', ['text_ngram'])             : 713 | ['__nlp__.10', '__nlp__.10 verbal', '__nlp__.10 verbal numeric', '__nlp__.2000', '__nlp__.2000 panel', ...]\n",
      "\t3.2s = Fit runtime\n",
      "\t116 features in original data used to generate 923 features in processed data.\n",
      "\tTrain Data (Processed) Memory Usage: 0.29 MB (0.0% of available memory)\n",
      "Data preprocessing and feature engineering runtime = 3.3s ...\n",
      "AutoGluon will gauge predictive performance using evaluation metric: 'accuracy'\n",
      "\tTo change this, specify the eval_metric parameter of Predictor()\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/learner.pkl\n",
      "Automatically generating train/validation split with holdout_frac=0.1, Train Rows: 108, Val Rows: 13\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/utils/data/X.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/utils/data/y.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/utils/data/X_val.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/utils/data/y_val.pkl\n",
      "Excluded Model Types: ['CAT', 'XGB']\n",
      "\tFound 'CAT' model in hyperparameters, but 'CAT' is present in `excluded_model_types` and will be removed.\n",
      "\tFound 'XGB' model in hyperparameters, but 'XGB' is present in `excluded_model_types` and will be removed.\n",
      "Fitting 11 L1 models ...\n",
      "Fitting model: KNeighborsUnif ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/KNeighborsUnif/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/utils/attr/KNeighborsUnif/y_pred_proba_val.pkl\n",
      "\t0.4615\t = Validation score   (accuracy)\n",
      "\t0.11s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/trainer.pkl\n",
      "Fitting model: KNeighborsDist ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/KNeighborsDist/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/utils/attr/KNeighborsDist/y_pred_proba_val.pkl\n",
      "\t0.4615\t = Validation score   (accuracy)\n",
      "\t0.11s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/trainer.pkl\n",
      "Fitting model: LightGBMXT ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/LightGBMXT/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/utils/attr/LightGBMXT/y_pred_proba_val.pkl\n",
      "\t0.8462\t = Validation score   (accuracy)\n",
      "\t0.78s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/trainer.pkl\n",
      "Fitting model: LightGBM ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/LightGBM/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/utils/attr/LightGBM/y_pred_proba_val.pkl\n",
      "\t0.8462\t = Validation score   (accuracy)\n",
      "\t0.59s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/trainer.pkl\n",
      "Fitting model: RandomForestGini ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/RandomForestGini/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/utils/attr/RandomForestGini/y_pred_proba_val.pkl\n",
      "\t0.8462\t = Validation score   (accuracy)\n",
      "\t1.16s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/trainer.pkl\n",
      "Fitting model: RandomForestEntr ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/RandomForestEntr/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/utils/attr/RandomForestEntr/y_pred_proba_val.pkl\n",
      "\t0.8462\t = Validation score   (accuracy)\n",
      "\t0.9s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/trainer.pkl\n",
      "Fitting model: ExtraTreesGini ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/ExtraTreesGini/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/utils/attr/ExtraTreesGini/y_pred_proba_val.pkl\n",
      "\t0.8462\t = Validation score   (accuracy)\n",
      "\t0.94s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/trainer.pkl\n",
      "Fitting model: ExtraTreesEntr ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/ExtraTreesEntr/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/utils/attr/ExtraTreesEntr/y_pred_proba_val.pkl\n",
      "\t0.8462\t = Validation score   (accuracy)\n",
      "\t0.9s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/trainer.pkl\n",
      "Fitting model: NeuralNetFastAI ...\n",
      "No improvement since epoch 1: early stopping\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/NeuralNetFastAI/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/NeuralNetFastAI/model-internals.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/utils/attr/NeuralNetFastAI/y_pred_proba_val.pkl\n",
      "\t0.9231\t = Validation score   (accuracy)\n",
      "\t1.14s\t = Training   runtime\n",
      "\t0.06s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/trainer.pkl\n",
      "Fitting model: NeuralNetTorch ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/NeuralNetTorch/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/utils/attr/NeuralNetTorch/y_pred_proba_val.pkl\n",
      "\t1.0\t = Validation score   (accuracy)\n",
      "\t0.6s\t = Training   runtime\n",
      "\t0.03s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/trainer.pkl\n",
      "Fitting model: LightGBMLarge ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/LightGBMLarge/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/utils/attr/LightGBMLarge/y_pred_proba_val.pkl\n",
      "\t0.8462\t = Validation score   (accuracy)\n",
      "\t1.05s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/trainer.pkl\n",
      "Fitting model: WeightedEnsemble_L2 ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/WeightedEnsemble_L2/utils/model_template.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/WeightedEnsemble_L2/utils/oof.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/WeightedEnsemble_L2/model.pkl\n",
      "\t1.0\t = Validation score   (accuracy)\n",
      "\t0.27s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/trainer.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/trainer.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/trainer.pkl\n",
      "AutoGluon training complete, total runtime = 19.15s ... Best model: \"WeightedEnsemble_L2\"\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/models/trainer.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/learner.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/predictor.pkl\n",
      "TabularPredictor saved. To load, use: predictor = TabularPredictor.load(\"s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-stroke_prediction/\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Training time= 19.940380811691284 seconds ---\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "start_time = time.time()\n",
    "buckt = sm_session.default_bucket()\n",
    "prefix= \"genomic-clinical-imaging-stroke_prediction\"\n",
    "save_file = 's3://{}/{}'.format(buckt, prefix)\n",
    "predictor = TabularPredictor(label= 'stroke_prediction', problem_type= 'binary', path=save_file).fit(train_data=training, holdout_frac=0.1, excluded_model_types=['CAT', 'XGB'])\n",
    "print(\"--- Training time= %s seconds ---\" % (time.time() - start_time))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "fe947544-7daa-4836-b74a-7291cb83d421",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Evaluation: accuracy on test data: 0.9666666666666667\n",
      "Evaluations on test data:\n",
      "{\n",
      "    \"accuracy\": 0.9666666666666667,\n",
      "    \"balanced_accuracy\": 0.9642857142857143,\n",
      "    \"mcc\": 0.9348527048856053,\n",
      "    \"f1\": 0.9696969696969697,\n",
      "    \"precision\": 0.9411764705882353,\n",
      "    \"recall\": 1.0\n",
      "}\n",
      "Detailed (per-class) classification report:\n",
      "{\n",
      "    \"0\": {\n",
      "        \"precision\": 1.0,\n",
      "        \"recall\": 0.9285714285714286,\n",
      "        \"f1-score\": 0.962962962962963,\n",
      "        \"support\": 14\n",
      "    },\n",
      "    \"1\": {\n",
      "        \"precision\": 0.9411764705882353,\n",
      "        \"recall\": 1.0,\n",
      "        \"f1-score\": 0.9696969696969697,\n",
      "        \"support\": 16\n",
      "    },\n",
      "    \"accuracy\": 0.9666666666666667,\n",
      "    \"macro avg\": {\n",
      "        \"precision\": 0.9705882352941176,\n",
      "        \"recall\": 0.9642857142857143,\n",
      "        \"f1-score\": 0.9663299663299664,\n",
      "        \"support\": 30\n",
      "    },\n",
      "    \"weighted avg\": {\n",
      "        \"precision\": 0.9686274509803922,\n",
      "        \"recall\": 0.9666666666666667,\n",
      "        \"f1-score\": 0.9665544332210999,\n",
      "        \"support\": 30\n",
      "    }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'accuracy': 0.9666666666666667,\n",
       " 'balanced_accuracy': 0.9642857142857143,\n",
       " 'mcc': 0.9348527048856053,\n",
       " 'f1': 0.9696969696969697,\n",
       " 'precision': 0.9411764705882353,\n",
       " 'recall': 1.0,\n",
       " 'confusion_matrix':     0   1\n",
       " 0  13   1\n",
       " 1   0  16,\n",
       " 'classification_report': {'0': {'precision': 1.0,\n",
       "   'recall': 0.9285714285714286,\n",
       "   'f1-score': 0.962962962962963,\n",
       "   'support': 14},\n",
       "  '1': {'precision': 0.9411764705882353,\n",
       "   'recall': 1.0,\n",
       "   'f1-score': 0.9696969696969697,\n",
       "   'support': 16},\n",
       "  'accuracy': 0.9666666666666667,\n",
       "  'macro avg': {'precision': 0.9705882352941176,\n",
       "   'recall': 0.9642857142857143,\n",
       "   'f1-score': 0.9663299663299664,\n",
       "   'support': 30},\n",
       "  'weighted avg': {'precision': 0.9686274509803922,\n",
       "   'recall': 0.9666666666666667,\n",
       "   'f1-score': 0.9665544332210999,\n",
       "   'support': 30}}}"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictor.evaluate_predictions(y_true=testing['stroke_prediction'], y_pred=predictor.predict(X_test), auxiliary_metrics=True, detailed_report=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cda76e3-db6c-44d3-a6e5-339a8344f55a",
   "metadata": {},
   "source": [
    "## Hypertension Prediction\n",
    "Splitting data for training and testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "926e139c-ec74-4c58-84c1-dca1c46f6022",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training size =  121\n",
      "Out of sample testing size =  30\n"
     ]
    }
   ],
   "source": [
    "#hypertension_prediction\n",
    "#Splitting data into training and testing 80:20\n",
    "training = dataset.sample(frac=0.8, random_state=25)\n",
    "training = training.drop(columns = ['patientid', 'alzheimers_prediction', 'coronary_heart_disease_prediction', 'stroke_prediction'])\n",
    "testing = dataset.drop(training.index)\n",
    "testing = testing.drop(columns = ['patientid', 'alzheimers_prediction', 'coronary_heart_disease_prediction', 'stroke_prediction'])\n",
    "X_test = testing.drop(columns = ['hypertension_prediction'])\n",
    "print(\"Training size = \", len(training))\n",
    "print(\"Out of sample testing size = \", len(testing))\n",
    "\n",
    "# Splitting data into training and testing for deploying to an endpoint\n",
    "training_target = training.pop(\"hypertension_prediction\")\n",
    "training.insert(0, 'hypertension_prediction', training_target)\n",
    "testing_target = testing.pop(\"hypertension_prediction\")\n",
    "testing.insert(0, 'hypertension_prediction', testing_target)\n",
    "training.to_csv('s3://multimodal-dataset-clinical-genomic-imaging/multimodal_hypertension_training.csv', index=False)\n",
    "testing.to_csv('s3://multimodal-dataset-clinical-genomic-imaging/multimodal_hypertension_testing.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6bf0905-5b45-4943-82a5-9211e40e010e",
   "metadata": {},
   "source": [
    "### Hypertension prediction on clinical, genomic, and imaging data using Autogluon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "ae669272-4ea4-4805-bb95-1e6dabf7fa4a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: path already exists! This predictor may overwrite an existing predictor! path=\"s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction\"\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/learner.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/predictor.pkl\n",
      "Beginning AutoGluon training ...\n",
      "AutoGluon will save models to \"s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/\"\n",
      "AutoGluon Version:  0.6.2\n",
      "Python Version:     3.7.10\n",
      "Operating System:   Linux\n",
      "Platform Machine:   x86_64\n",
      "Platform Version:   #1 SMP Thu Mar 23 09:54:12 UTC 2023\n",
      "Train Data Rows:    121\n",
      "Train Data Columns: 144\n",
      "Label Column: hypertension_prediction\n",
      "Preprocessing data ...\n",
      "Selected class <--> label mapping:  class 1 = 1, class 0 = 0\n",
      "Using Feature Generators to preprocess the data ...\n",
      "Fitting AutoMLPipelineFeatureGenerator...\n",
      "\tAvailable Memory:                    2779.76 MB\n",
      "\tTrain Data (Original)  Memory Usage: 1.52 MB (0.1% of available memory)\n",
      "\tInferring data type of each feature based on column values. Set feature_metadata_in to manually specify special dtypes of the features.\n",
      "\tStage 1 Generators:\n",
      "\t\tFitting AsTypeFeatureGenerator...\n",
      "\t\t\tNote: Converting 8 features to boolean dtype as they only contain 2 unique values.\n",
      "\tStage 2 Generators:\n",
      "\t\tFitting FillNaFeatureGenerator...\n",
      "\tStage 3 Generators:\n",
      "\t\tFitting IdentityFeatureGenerator...\n",
      "\t\tFitting CategoryFeatureGenerator...\n",
      "\t\t\tFitting CategoryMemoryMinimizeFeatureGenerator...\n",
      "\t\tFitting DatetimeFeatureGenerator...\n",
      "/opt/conda/lib/python3.7/site-packages/autogluon/features/generators/datetime.py:59: FutureWarning: casting datetime64[ns, UTC] values to int64 with .astype(...) is deprecated and will raise in a future version. Use .view(...) instead.\n",
      "  good_rows = series[~series.isin(bad_rows)].astype(np.int64)\n",
      "\t\tFitting TextSpecialFeatureGenerator...\n",
      "\t\t\tFitting BinnedFeatureGenerator...\n",
      "\t\t\tFitting DropDuplicatesFeatureGenerator...\n",
      "\t\tFitting TextNgramFeatureGenerator...\n",
      "\t\t\tFitting CountVectorizer for text features: ['diagnostictype', 'diagnosticresult', 'conditionid', 'vaccinations', 'imagingname', 'imagebodypart', 'imagetitle', 'encounterreason', 'observationtype', 'observationtext', 'observationdisplay', 'proceduretype']\n",
      "\t\t\tCountVectorizer fit with vocabulary size = 1161\n",
      "\t\tWarning: Due to memory constraints, ngram feature count is being reduced. Allocate more memory to maximize model quality.\n",
      "\t\tReducing Vectorizer vocab size from 1161 to 726 to avoid OOM error\n",
      "\tStage 4 Generators:\n",
      "\t\tFitting DropUniqueFeatureGenerator...\n",
      "\tUseless Original Features (Count: 27): ['clinical_significance', 'contigname', 'referenceallele', 'alternatealleles', 'phased', 'deceasedindicator', 'encounterdiagnosis', 'hospitalizationstatus', 'medicationstatus', 'claim', 'original_shape_elongation', 'original_shape_flatness', 'original_shape_leastaxislength', 'original_shape_majoraxislength', 'original_shape_maximum2ddiametercolumn', 'original_shape_maximum2ddiameterrow', 'original_shape_maximum2ddiameterslice', 'original_shape_maximum3ddiameter', 'original_shape_meshvolume', 'original_shape_minoraxislength', 'original_shape_sphericity', 'original_shape_surfacearea', 'original_shape_surfacevolumeratio', 'original_shape_voxelvolume', 'original_firstorder_maximum', 'original_firstorder_minimum', 'original_firstorder_range']\n",
      "\t\tThese features carry no predictive signal and should be manually investigated.\n",
      "\t\tThis is typically a feature which has the same value for all rows.\n",
      "\t\tThese features do not need to be present at inference time.\n",
      "\tUnused Original Features (Count: 1): ['imagesetid']\n",
      "\t\tThese features were not used to generate any of the output features. Add a feature generator compatible with these features to utilize them.\n",
      "\t\tFeatures can also be unused if they carry very little information, such as being categorical but having almost entirely unique values or being duplicates of other features.\n",
      "\t\tThese features do not need to be present at inference time.\n",
      "\t\t('object', []) : 1 | ['imagesetid']\n",
      "\tTypes of features in original data (raw dtype, special dtypes):\n",
      "\t\t('float', [])                      : 91 | ['postalcode', 'original_firstorder_10percentile', 'original_firstorder_90percentile', 'original_firstorder_energy', 'original_firstorder_entropy', ...]\n",
      "\t\t('object', [])                     :  6 | ['gender', 'maritalstatus', 'categorycode', 'categorydescription', 'imagecode', ...]\n",
      "\t\t('object', ['datetime_as_object']) :  2 | ['deceaseddatetime', 'medicationdate']\n",
      "\t\t('object', ['text'])               : 17 | ['gene_info', 'start', 'calls', 'diagnostictype', 'diagnosticcategory', ...]\n",
      "\tTypes of features in processed data (raw dtype, special dtypes):\n",
      "\t\t('category', [])                    :   3 | ['gender', 'maritalstatus', 'imagecode']\n",
      "\t\t('category', ['text_as_category'])  :  11 | ['diagnostictype', 'diagnosticresult', 'conditionid', 'vaccinations', 'imagingname', ...]\n",
      "\t\t('float', [])                       :  91 | ['postalcode', 'original_firstorder_10percentile', 'original_firstorder_90percentile', 'original_firstorder_energy', 'original_firstorder_entropy', ...]\n",
      "\t\t('int', ['binned', 'text_special']) :  92 | ['diagnostictype.char_count', 'diagnostictype.word_count', 'diagnostictype.capital_ratio', 'diagnostictype.lower_ratio', 'diagnostictype.digit_ratio', ...]\n",
      "\t\t('int', ['bool'])                   :   8 | ['gene_info', 'start', 'calls', 'diagnosticcategory', 'categorycode', ...]\n",
      "\t\t('int', ['datetime_as_int'])        :  10 | ['deceaseddatetime', 'deceaseddatetime.year', 'deceaseddatetime.month', 'deceaseddatetime.day', 'deceaseddatetime.dayofweek', ...]\n",
      "\t\t('int', ['text_ngram'])             : 717 | ['__nlp__.10', '__nlp__.10 verbal', '__nlp__.10 verbal numeric', '__nlp__.2000', '__nlp__.2000 panel', ...]\n",
      "\t3.2s = Fit runtime\n",
      "\t116 features in original data used to generate 932 features in processed data.\n",
      "\tTrain Data (Processed) Memory Usage: 0.29 MB (0.0% of available memory)\n",
      "Data preprocessing and feature engineering runtime = 3.37s ...\n",
      "AutoGluon will gauge predictive performance using evaluation metric: 'accuracy'\n",
      "\tTo change this, specify the eval_metric parameter of Predictor()\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/learner.pkl\n",
      "Automatically generating train/validation split with holdout_frac=0.1, Train Rows: 108, Val Rows: 13\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/utils/data/X.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/utils/data/y.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/utils/data/X_val.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/utils/data/y_val.pkl\n",
      "Excluded Model Types: ['CAT', 'XGB']\n",
      "\tFound 'CAT' model in hyperparameters, but 'CAT' is present in `excluded_model_types` and will be removed.\n",
      "\tFound 'XGB' model in hyperparameters, but 'XGB' is present in `excluded_model_types` and will be removed.\n",
      "Fitting 11 L1 models ...\n",
      "Fitting model: KNeighborsUnif ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/KNeighborsUnif/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/utils/attr/KNeighborsUnif/y_pred_proba_val.pkl\n",
      "\t0.6923\t = Validation score   (accuracy)\n",
      "\t0.1s\t = Training   runtime\n",
      "\t0.1s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/trainer.pkl\n",
      "Fitting model: KNeighborsDist ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/KNeighborsDist/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/utils/attr/KNeighborsDist/y_pred_proba_val.pkl\n",
      "\t0.6923\t = Validation score   (accuracy)\n",
      "\t0.12s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/trainer.pkl\n",
      "Fitting model: LightGBMXT ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/LightGBMXT/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/utils/attr/LightGBMXT/y_pred_proba_val.pkl\n",
      "\t0.8462\t = Validation score   (accuracy)\n",
      "\t0.69s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/trainer.pkl\n",
      "Fitting model: LightGBM ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/LightGBM/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/utils/attr/LightGBM/y_pred_proba_val.pkl\n",
      "\t0.8462\t = Validation score   (accuracy)\n",
      "\t0.65s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/trainer.pkl\n",
      "Fitting model: RandomForestGini ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/RandomForestGini/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/utils/attr/RandomForestGini/y_pred_proba_val.pkl\n",
      "\t0.7692\t = Validation score   (accuracy)\n",
      "\t0.88s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/trainer.pkl\n",
      "Fitting model: RandomForestEntr ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/RandomForestEntr/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/utils/attr/RandomForestEntr/y_pred_proba_val.pkl\n",
      "\t0.7692\t = Validation score   (accuracy)\n",
      "\t0.88s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/trainer.pkl\n",
      "Fitting model: ExtraTreesGini ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/ExtraTreesGini/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/utils/attr/ExtraTreesGini/y_pred_proba_val.pkl\n",
      "\t0.6923\t = Validation score   (accuracy)\n",
      "\t1.26s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/trainer.pkl\n",
      "Fitting model: ExtraTreesEntr ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/ExtraTreesEntr/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/utils/attr/ExtraTreesEntr/y_pred_proba_val.pkl\n",
      "\t0.6923\t = Validation score   (accuracy)\n",
      "\t0.92s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/trainer.pkl\n",
      "Fitting model: NeuralNetFastAI ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/NeuralNetFastAI/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/NeuralNetFastAI/model-internals.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/utils/attr/NeuralNetFastAI/y_pred_proba_val.pkl\n",
      "\t0.9231\t = Validation score   (accuracy)\n",
      "\t1.59s\t = Training   runtime\n",
      "\t0.07s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/trainer.pkl\n",
      "Fitting model: NeuralNetTorch ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/NeuralNetTorch/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/utils/attr/NeuralNetTorch/y_pred_proba_val.pkl\n",
      "\t0.9231\t = Validation score   (accuracy)\n",
      "\t1.31s\t = Training   runtime\n",
      "\t0.03s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/trainer.pkl\n",
      "Fitting model: LightGBMLarge ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/LightGBMLarge/model.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/utils/attr/LightGBMLarge/y_pred_proba_val.pkl\n",
      "\t0.8462\t = Validation score   (accuracy)\n",
      "\t1.18s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/trainer.pkl\n",
      "Fitting model: WeightedEnsemble_L2 ...\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/WeightedEnsemble_L2/utils/model_template.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/WeightedEnsemble_L2/utils/oof.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/WeightedEnsemble_L2/model.pkl\n",
      "\t0.9231\t = Validation score   (accuracy)\n",
      "\t0.28s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/trainer.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/trainer.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/trainer.pkl\n",
      "AutoGluon training complete, total runtime = 20.94s ... Best model: \"WeightedEnsemble_L2\"\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/models/trainer.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/learner.pkl\n",
      "save object to s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/predictor.pkl\n",
      "TabularPredictor saved. To load, use: predictor = TabularPredictor.load(\"s3://sagemaker-us-east-1-659535263284/genomic-clinical-imaging-hypertension-prediction/\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Training time= 21.77889585494995 seconds ---\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "start_time = time.time()\n",
    "buckt = sm_session.default_bucket()\n",
    "prefix= \"genomic-clinical-imaging-hypertension-prediction\"\n",
    "save_file = 's3://{}/{}'.format(buckt, prefix)\n",
    "predictor = TabularPredictor(label= 'hypertension_prediction', problem_type= 'binary', path=save_file).fit(train_data=training, holdout_frac=0.1, excluded_model_types=['CAT', 'XGB'])\n",
    "print(\"--- Training time= %s seconds ---\" % (time.time() - start_time))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "1611732f-6432-4713-b8da-90cbf404d3ec",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Evaluation: accuracy on test data: 0.8666666666666667\n",
      "Evaluations on test data:\n",
      "{\n",
      "    \"accuracy\": 0.8666666666666667,\n",
      "    \"balanced_accuracy\": 0.9047619047619048,\n",
      "    \"mcc\": 0.7486251134176306,\n",
      "    \"f1\": 0.8181818181818181,\n",
      "    \"precision\": 0.6923076923076923,\n",
      "    \"recall\": 1.0\n",
      "}\n",
      "Detailed (per-class) classification report:\n",
      "{\n",
      "    \"0\": {\n",
      "        \"precision\": 1.0,\n",
      "        \"recall\": 0.8095238095238095,\n",
      "        \"f1-score\": 0.8947368421052632,\n",
      "        \"support\": 21\n",
      "    },\n",
      "    \"1\": {\n",
      "        \"precision\": 0.6923076923076923,\n",
      "        \"recall\": 1.0,\n",
      "        \"f1-score\": 0.8181818181818181,\n",
      "        \"support\": 9\n",
      "    },\n",
      "    \"accuracy\": 0.8666666666666667,\n",
      "    \"macro avg\": {\n",
      "        \"precision\": 0.8461538461538461,\n",
      "        \"recall\": 0.9047619047619048,\n",
      "        \"f1-score\": 0.8564593301435406,\n",
      "        \"support\": 30\n",
      "    },\n",
      "    \"weighted avg\": {\n",
      "        \"precision\": 0.9076923076923077,\n",
      "        \"recall\": 0.8666666666666667,\n",
      "        \"f1-score\": 0.8717703349282296,\n",
      "        \"support\": 30\n",
      "    }\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'accuracy': 0.8666666666666667,\n",
       " 'balanced_accuracy': 0.9047619047619048,\n",
       " 'mcc': 0.7486251134176306,\n",
       " 'f1': 0.8181818181818181,\n",
       " 'precision': 0.6923076923076923,\n",
       " 'recall': 1.0,\n",
       " 'confusion_matrix':     0  1\n",
       " 0  17  4\n",
       " 1   0  9,\n",
       " 'classification_report': {'0': {'precision': 1.0,\n",
       "   'recall': 0.8095238095238095,\n",
       "   'f1-score': 0.8947368421052632,\n",
       "   'support': 21},\n",
       "  '1': {'precision': 0.6923076923076923,\n",
       "   'recall': 1.0,\n",
       "   'f1-score': 0.8181818181818181,\n",
       "   'support': 9},\n",
       "  'accuracy': 0.8666666666666667,\n",
       "  'macro avg': {'precision': 0.8461538461538461,\n",
       "   'recall': 0.9047619047619048,\n",
       "   'f1-score': 0.8564593301435406,\n",
       "   'support': 30},\n",
       "  'weighted avg': {'precision': 0.9076923076923077,\n",
       "   'recall': 0.8666666666666667,\n",
       "   'f1-score': 0.8717703349282296,\n",
       "   'support': 30}}}"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictor.evaluate_predictions(y_true=testing['hypertension_prediction'], y_pred=predictor.predict(X_test), auxiliary_metrics=True, detailed_report=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6669950-15c3-4231-b10d-69971fdd656a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "availableInstances": [
   {
    "_defaultOrder": 0,
    "_isFastLaunch": true,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 4,
    "name": "ml.t3.medium",
    "vcpuNum": 2
   },
   {
    "_defaultOrder": 1,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 8,
    "name": "ml.t3.large",
    "vcpuNum": 2
   },
   {
    "_defaultOrder": 2,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 16,
    "name": "ml.t3.xlarge",
    "vcpuNum": 4
   },
   {
    "_defaultOrder": 3,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 32,
    "name": "ml.t3.2xlarge",
    "vcpuNum": 8
   },
   {
    "_defaultOrder": 4,
    "_isFastLaunch": true,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 8,
    "name": "ml.m5.large",
    "vcpuNum": 2
   },
   {
    "_defaultOrder": 5,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 16,
    "name": "ml.m5.xlarge",
    "vcpuNum": 4
   },
   {
    "_defaultOrder": 6,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 32,
    "name": "ml.m5.2xlarge",
    "vcpuNum": 8
   },
   {
    "_defaultOrder": 7,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 64,
    "name": "ml.m5.4xlarge",
    "vcpuNum": 16
   },
   {
    "_defaultOrder": 8,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 128,
    "name": "ml.m5.8xlarge",
    "vcpuNum": 32
   },
   {
    "_defaultOrder": 9,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 192,
    "name": "ml.m5.12xlarge",
    "vcpuNum": 48
   },
   {
    "_defaultOrder": 10,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 256,
    "name": "ml.m5.16xlarge",
    "vcpuNum": 64
   },
   {
    "_defaultOrder": 11,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 384,
    "name": "ml.m5.24xlarge",
    "vcpuNum": 96
   },
   {
    "_defaultOrder": 12,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 8,
    "name": "ml.m5d.large",
    "vcpuNum": 2
   },
   {
    "_defaultOrder": 13,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 16,
    "name": "ml.m5d.xlarge",
    "vcpuNum": 4
   },
   {
    "_defaultOrder": 14,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 32,
    "name": "ml.m5d.2xlarge",
    "vcpuNum": 8
   },
   {
    "_defaultOrder": 15,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 64,
    "name": "ml.m5d.4xlarge",
    "vcpuNum": 16
   },
   {
    "_defaultOrder": 16,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 128,
    "name": "ml.m5d.8xlarge",
    "vcpuNum": 32
   },
   {
    "_defaultOrder": 17,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 192,
    "name": "ml.m5d.12xlarge",
    "vcpuNum": 48
   },
   {
    "_defaultOrder": 18,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 256,
    "name": "ml.m5d.16xlarge",
    "vcpuNum": 64
   },
   {
    "_defaultOrder": 19,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 384,
    "name": "ml.m5d.24xlarge",
    "vcpuNum": 96
   },
   {
    "_defaultOrder": 20,
    "_isFastLaunch": false,
    "category": "General purpose",
    "gpuNum": 0,
    "hideHardwareSpecs": true,
    "memoryGiB": 0,
    "name": "ml.geospatial.interactive",
    "supportedImageNames": [
     "sagemaker-geospatial-v1-0"
    ],
    "vcpuNum": 0
   },
   {
    "_defaultOrder": 21,
    "_isFastLaunch": true,
    "category": "Compute optimized",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 4,
    "name": "ml.c5.large",
    "vcpuNum": 2
   },
   {
    "_defaultOrder": 22,
    "_isFastLaunch": false,
    "category": "Compute optimized",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 8,
    "name": "ml.c5.xlarge",
    "vcpuNum": 4
   },
   {
    "_defaultOrder": 23,
    "_isFastLaunch": false,
    "category": "Compute optimized",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 16,
    "name": "ml.c5.2xlarge",
    "vcpuNum": 8
   },
   {
    "_defaultOrder": 24,
    "_isFastLaunch": false,
    "category": "Compute optimized",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 32,
    "name": "ml.c5.4xlarge",
    "vcpuNum": 16
   },
   {
    "_defaultOrder": 25,
    "_isFastLaunch": false,
    "category": "Compute optimized",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 72,
    "name": "ml.c5.9xlarge",
    "vcpuNum": 36
   },
   {
    "_defaultOrder": 26,
    "_isFastLaunch": false,
    "category": "Compute optimized",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 96,
    "name": "ml.c5.12xlarge",
    "vcpuNum": 48
   },
   {
    "_defaultOrder": 27,
    "_isFastLaunch": false,
    "category": "Compute optimized",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 144,
    "name": "ml.c5.18xlarge",
    "vcpuNum": 72
   },
   {
    "_defaultOrder": 28,
    "_isFastLaunch": false,
    "category": "Compute optimized",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 192,
    "name": "ml.c5.24xlarge",
    "vcpuNum": 96
   },
   {
    "_defaultOrder": 29,
    "_isFastLaunch": true,
    "category": "Accelerated computing",
    "gpuNum": 1,
    "hideHardwareSpecs": false,
    "memoryGiB": 16,
    "name": "ml.g4dn.xlarge",
    "vcpuNum": 4
   },
   {
    "_defaultOrder": 30,
    "_isFastLaunch": false,
    "category": "Accelerated computing",
    "gpuNum": 1,
    "hideHardwareSpecs": false,
    "memoryGiB": 32,
    "name": "ml.g4dn.2xlarge",
    "vcpuNum": 8
   },
   {
    "_defaultOrder": 31,
    "_isFastLaunch": false,
    "category": "Accelerated computing",
    "gpuNum": 1,
    "hideHardwareSpecs": false,
    "memoryGiB": 64,
    "name": "ml.g4dn.4xlarge",
    "vcpuNum": 16
   },
   {
    "_defaultOrder": 32,
    "_isFastLaunch": false,
    "category": "Accelerated computing",
    "gpuNum": 1,
    "hideHardwareSpecs": false,
    "memoryGiB": 128,
    "name": "ml.g4dn.8xlarge",
    "vcpuNum": 32
   },
   {
    "_defaultOrder": 33,
    "_isFastLaunch": false,
    "category": "Accelerated computing",
    "gpuNum": 4,
    "hideHardwareSpecs": false,
    "memoryGiB": 192,
    "name": "ml.g4dn.12xlarge",
    "vcpuNum": 48
   },
   {
    "_defaultOrder": 34,
    "_isFastLaunch": false,
    "category": "Accelerated computing",
    "gpuNum": 1,
    "hideHardwareSpecs": false,
    "memoryGiB": 256,
    "name": "ml.g4dn.16xlarge",
    "vcpuNum": 64
   },
   {
    "_defaultOrder": 35,
    "_isFastLaunch": false,
    "category": "Accelerated computing",
    "gpuNum": 1,
    "hideHardwareSpecs": false,
    "memoryGiB": 61,
    "name": "ml.p3.2xlarge",
    "vcpuNum": 8
   },
   {
    "_defaultOrder": 36,
    "_isFastLaunch": false,
    "category": "Accelerated computing",
    "gpuNum": 4,
    "hideHardwareSpecs": false,
    "memoryGiB": 244,
    "name": "ml.p3.8xlarge",
    "vcpuNum": 32
   },
   {
    "_defaultOrder": 37,
    "_isFastLaunch": false,
    "category": "Accelerated computing",
    "gpuNum": 8,
    "hideHardwareSpecs": false,
    "memoryGiB": 488,
    "name": "ml.p3.16xlarge",
    "vcpuNum": 64
   },
   {
    "_defaultOrder": 38,
    "_isFastLaunch": false,
    "category": "Accelerated computing",
    "gpuNum": 8,
    "hideHardwareSpecs": false,
    "memoryGiB": 768,
    "name": "ml.p3dn.24xlarge",
    "vcpuNum": 96
   },
   {
    "_defaultOrder": 39,
    "_isFastLaunch": false,
    "category": "Memory Optimized",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 16,
    "name": "ml.r5.large",
    "vcpuNum": 2
   },
   {
    "_defaultOrder": 40,
    "_isFastLaunch": false,
    "category": "Memory Optimized",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 32,
    "name": "ml.r5.xlarge",
    "vcpuNum": 4
   },
   {
    "_defaultOrder": 41,
    "_isFastLaunch": false,
    "category": "Memory Optimized",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 64,
    "name": "ml.r5.2xlarge",
    "vcpuNum": 8
   },
   {
    "_defaultOrder": 42,
    "_isFastLaunch": false,
    "category": "Memory Optimized",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 128,
    "name": "ml.r5.4xlarge",
    "vcpuNum": 16
   },
   {
    "_defaultOrder": 43,
    "_isFastLaunch": false,
    "category": "Memory Optimized",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 256,
    "name": "ml.r5.8xlarge",
    "vcpuNum": 32
   },
   {
    "_defaultOrder": 44,
    "_isFastLaunch": false,
    "category": "Memory Optimized",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 384,
    "name": "ml.r5.12xlarge",
    "vcpuNum": 48
   },
   {
    "_defaultOrder": 45,
    "_isFastLaunch": false,
    "category": "Memory Optimized",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 512,
    "name": "ml.r5.16xlarge",
    "vcpuNum": 64
   },
   {
    "_defaultOrder": 46,
    "_isFastLaunch": false,
    "category": "Memory Optimized",
    "gpuNum": 0,
    "hideHardwareSpecs": false,
    "memoryGiB": 768,
    "name": "ml.r5.24xlarge",
    "vcpuNum": 96
   },
   {
    "_defaultOrder": 47,
    "_isFastLaunch": false,
    "category": "Accelerated computing",
    "gpuNum": 1,
    "hideHardwareSpecs": false,
    "memoryGiB": 16,
    "name": "ml.g5.xlarge",
    "vcpuNum": 4
   },
   {
    "_defaultOrder": 48,
    "_isFastLaunch": false,
    "category": "Accelerated computing",
    "gpuNum": 1,
    "hideHardwareSpecs": false,
    "memoryGiB": 32,
    "name": "ml.g5.2xlarge",
    "vcpuNum": 8
   },
   {
    "_defaultOrder": 49,
    "_isFastLaunch": false,
    "category": "Accelerated computing",
    "gpuNum": 1,
    "hideHardwareSpecs": false,
    "memoryGiB": 64,
    "name": "ml.g5.4xlarge",
    "vcpuNum": 16
   },
   {
    "_defaultOrder": 50,
    "_isFastLaunch": false,
    "category": "Accelerated computing",
    "gpuNum": 1,
    "hideHardwareSpecs": false,
    "memoryGiB": 128,
    "name": "ml.g5.8xlarge",
    "vcpuNum": 32
   },
   {
    "_defaultOrder": 51,
    "_isFastLaunch": false,
    "category": "Accelerated computing",
    "gpuNum": 1,
    "hideHardwareSpecs": false,
    "memoryGiB": 256,
    "name": "ml.g5.16xlarge",
    "vcpuNum": 64
   },
   {
    "_defaultOrder": 52,
    "_isFastLaunch": false,
    "category": "Accelerated computing",
    "gpuNum": 4,
    "hideHardwareSpecs": false,
    "memoryGiB": 192,
    "name": "ml.g5.12xlarge",
    "vcpuNum": 48
   },
   {
    "_defaultOrder": 53,
    "_isFastLaunch": false,
    "category": "Accelerated computing",
    "gpuNum": 4,
    "hideHardwareSpecs": false,
    "memoryGiB": 384,
    "name": "ml.g5.24xlarge",
    "vcpuNum": 96
   },
   {
    "_defaultOrder": 54,
    "_isFastLaunch": false,
    "category": "Accelerated computing",
    "gpuNum": 8,
    "hideHardwareSpecs": false,
    "memoryGiB": 768,
    "name": "ml.g5.48xlarge",
    "vcpuNum": 192
   }
  ],
  "instance_type": "ml.t3.medium",
  "kernelspec": {
   "display_name": "conda_mxnet_p38",
   "language": "python",
   "name": "conda_mxnet_p38"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
