{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f9472cd9",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3c8ef7a7",
   "metadata": {},
   "source": [
    "# Create Amazon Omics Analytic Stores to Import Genomic Data and Run Queries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85d3918d",
   "metadata": {},
   "source": [
    "### Follow steps in this notebook to: 1) create Amazon Omics Reference, Variant, and Annotation Stores; 2) import reference genome, variant files, and ClinVar annotation file from S3 to the respective data stores; 3) query the variant and annotation data. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aef34065",
   "metadata": {},
   "source": [
    "## Create reference store and import reference genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2dee102e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "import itertools as it\n",
    "import json\n",
    "import gzip\n",
    "import os\n",
    "from pprint import pprint\n",
    "import time\n",
    "import urllib\n",
    "\n",
    "import boto3\n",
    "import botocore.exceptions\n",
    "import requests"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86a19a6d",
   "metadata": {},
   "source": [
    "### Create service role"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a036d0a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set a timestamp\n",
    "dt_fmt = '%Y%m%dT%H%M%S'\n",
    "ts = datetime.now().strftime(dt_fmt)\n",
    "\n",
    "policy = {\n",
    "  \"Version\": \"2012-10-17\",\n",
    "  \"Statement\": [\n",
    "    {\n",
    "      \"Effect\": \"Allow\",\n",
    "      \"Action\": [\n",
    "        \"omics:*\"\n",
    "      ],\n",
    "      \"Resource\": \"*\"\n",
    "    },\n",
    "    {\n",
    "      \"Effect\": \"Allow\",\n",
    "      \"Action\": [\n",
    "        \"ram:AcceptResourceShareInvitation\",\n",
    "        \"ram:GetResourceShareInvitations\"\n",
    "      ],\n",
    "      \"Resource\": \"*\"\n",
    "    },\n",
    "    {\n",
    "      \"Effect\": \"Allow\",\n",
    "      \"Action\": [\n",
    "        \"s3:GetBucketLocation\",\n",
    "        \"s3:PutObject\",\n",
    "        \"s3:GetObject\",\n",
    "        \"s3:ListBucket\",\n",
    "        \"s3:AbortMultipartUpload\",\n",
    "        \"s3:ListMultipartUploadParts\",\n",
    "        \"s3:GetObjectAcl\",\n",
    "        \"s3:PutObjectAcl\"\n",
    "      ],\n",
    "      \"Resource\": \"*\"\n",
    "    }\n",
    "  ]\n",
    "}\n",
    "\n",
    "trust_policy = {\n",
    "    \"Version\": \"2012-10-17\",\n",
    "    \"Statement\": [\n",
    "        {\n",
    "            \"Effect\": \"Allow\",\n",
    "            \"Principal\": {\n",
    "                \"Service\": \"omics.amazonaws.com\"\n",
    "            },\n",
    "            \"Action\": \"sts:AssumeRole\"\n",
    "        }\n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9eb174a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Base name for role and policy\n",
    "omics_iam_name = f'multimodal-omics-{ts}'\n",
    "\n",
    "# Create the IAM client\n",
    "iam = boto3.resource('iam')\n",
    "\n",
    "# Check if the role already exist. If not, create it\n",
    "try:\n",
    "    role = iam.Role(omics_iam_name)\n",
    "    role.load()\n",
    "    \n",
    "except botocore.exceptions.ClientError as ex:\n",
    "    if ex.response[\"Error\"][\"Code\"] == \"NoSuchEntity\":\n",
    "        #Create the role with the corresponding trust policy\n",
    "        role = iam.create_role(\n",
    "            RoleName=omics_iam_name, \n",
    "            AssumeRolePolicyDocument=json.dumps(trust_policy))\n",
    "        \n",
    "        #Create policy\n",
    "        policy = iam.create_policy(\n",
    "            PolicyName='{}-policy'.format(omics_iam_name), \n",
    "            Description=\"Policy for Amazon Omics\",\n",
    "            PolicyDocument=json.dumps(policy))\n",
    "        \n",
    "        #Attach the policy to the role\n",
    "        policy.attach_role(RoleName=omics_iam_name)\n",
    "    else:\n",
    "        print (ex)\n",
    "        print('Somthing went wrong, please retry and check your account settings and permissions')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c0a3e04a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_role_arn(role_name):\n",
    "    try:\n",
    "        iam = boto3.resource('iam')\n",
    "        role = iam.Role(role_name)\n",
    "        role.load()  # calls GetRole to load attributes\n",
    "    except botocore.exeptions.ClientError:\n",
    "        print(\"Couldn't get role named %s.\"%role_name)\n",
    "        raise\n",
    "    else:\n",
    "        return role.arn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "161d02b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace <S3 URI for FASTA file> with the S3 URI of the reference genome in FASTA format. \n",
    "\n",
    "SOURCE_S3_URIS = {\n",
    "    \"reference\": \"<S3 URI for FASTA file>\"\n",
    "    }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7243a7e",
   "metadata": {},
   "source": [
    "### Create Omics client"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c02fcf1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "omics = boto3.client('omics', region_name='us-east-1')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec7c48f4",
   "metadata": {},
   "source": [
    "### Create reference store "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1657476d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ref_store_id(client=None):\n",
    "    if not client:\n",
    "        client = boto3.client('omics')\n",
    "    \n",
    "    resp = client.list_reference_stores(maxResults=10)\n",
    "    list_of_stores = resp.get('referenceStores')\n",
    "    store_id = None\n",
    "    \n",
    "    if list_of_stores != None:\n",
    "        # Since there can only be one store per region, if there is a store present use the first one\n",
    "        store_id = list_of_stores[0].get('id')\n",
    "    \n",
    "    return store_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3071ab4",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Checking for a reference store in region: {omics.meta.region_name}\")\n",
    "if get_ref_store_id(omics) == None:\n",
    "    response = omics.create_reference_store(name='myReferenceStore')\n",
    "    print(response)\n",
    "else:\n",
    "    print(\"Congratulations, you have an existing reference store!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9aee28d4",
   "metadata": {},
   "source": [
    "### Import reference genome to reference store"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6bf75b4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If using a different reference genomem, replace \"hg19\" with a different prefix\n",
    "\n",
    "ref_name = f'hg19-{ts}'\n",
    "\n",
    "ref_import_job = omics.start_reference_import_job(\n",
    "    referenceStoreId=get_ref_store_id(omics), \n",
    "    roleArn=get_role_arn(omics_iam_name),\n",
    "    sources=[{\n",
    "        'sourceFile': SOURCE_S3_URIS[\"reference\"],\n",
    "        'name': ref_name,\n",
    "        'tags': {'SourceLocation': '1kg'}\n",
    "    }])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "219c738a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_import_job = omics.get_reference_import_job(\n",
    "    referenceStoreId=get_ref_store_id(omics), \n",
    "    id=ref_import_job['id'])\n",
    "ref_import_job"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0f91c51",
   "metadata": {},
   "source": [
    "### !!! Wait until the above import job has finished !!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "055d8c70",
   "metadata": {},
   "outputs": [],
   "source": [
    "resp = omics.list_references(referenceStoreId=get_ref_store_id(omics), filter={\"name\": ref_name})\n",
    "\n",
    "ref_list = resp\n",
    "pprint(resp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41525f3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store this reference\n",
    "ref = omics.get_reference_metadata(\n",
    "    referenceStoreId=get_ref_store_id(omics), \n",
    "    id=ref_list['references'][0]['id'])\n",
    "ref"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9ed9b75",
   "metadata": {},
   "source": [
    "## Create Variant Store and import VCF files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "b0796c2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_reference_arn(ref_name, client=None):\n",
    "    if not client:\n",
    "        client = boto3.client('omics')\n",
    "    \n",
    "    resp = client.list_reference_stores(maxResults=10)\n",
    "    ref_stores = resp.get('referenceStores')\n",
    "    \n",
    "    # There can only be one reference store per account per region\n",
    "    # if there is a store present, it is the first one\n",
    "    ref_store = ref_stores[0] if ref_stores else None\n",
    "    \n",
    "    if not ref_store:\n",
    "        raise RuntimeError(\"You have not created a reference store, please got to the Amazon Omics Storage tutorial to learn how to create one. Do not continue with this notebook\")\n",
    "        \n",
    "    ref_arn = None\n",
    "    resp = omics.list_references(referenceStoreId=ref_store['id'])\n",
    "    ref_list = resp.get('references')\n",
    "    \n",
    "    for ref in resp.get('references'):\n",
    "        if ref['name'] == ref_name:\n",
    "            ref_arn = ref['arn']\n",
    "    \n",
    "    if ref_arn == None:\n",
    "        raise RuntimeError(f\"Could not find {ref_name}.\")\n",
    "    \n",
    "    return ref_arn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "01ec4e3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace <S3 URI for VCF files> with the S3 URI of the bucket containing the VCF files. \n",
    "\n",
    "SOURCE_VARIANT_URI = \"<S3 URI for VCF files>\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "801a2d03",
   "metadata": {},
   "outputs": [],
   "source": [
    "source = urllib.parse.urlparse(SOURCE_VARIANT_URI)\n",
    "bucket = source.netloc\n",
    "prefix = source.path[1:]\n",
    "\n",
    "s3r = boto3.resource('s3')\n",
    "\n",
    "bucket = s3r.Bucket(bucket)\n",
    "objects = bucket.objects.filter(Prefix=prefix, MaxKeys=10_000)\n",
    "ext = '.vcf'\n",
    "\n",
    "vcf_list = [f\"s3://{o.bucket_name}/{o.key}\" for o in objects if o.key.endswith(ext)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff428d41",
   "metadata": {},
   "source": [
    "### Create Variant Store"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c1fa8fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_store_name = f'synthea_newvariants_{ts.lower()}'\n",
    "\n",
    "# Replace <reference name> with reference genome name created in the previous steps\n",
    "ref_name = '<reference name>' \n",
    "\n",
    "response = omics.create_variant_store(\n",
    "    name=var_store_name, \n",
    "    reference={\"referenceArn\": get_reference_arn(ref_name, omics)}\n",
    ")\n",
    "\n",
    "var_store = response\n",
    "response"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82351bce",
   "metadata": {},
   "source": [
    "### !!! Wait until the Variant Store is created !!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3aed46b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    waiter = omics.get_waiter('variant_store_created')\n",
    "    waiter.wait(name=var_store['name'])\n",
    "\n",
    "    print(f\"variant store {var_store['name']} ready for use\")\n",
    "except botocore.exceptions.WaiterError as e:\n",
    "    print(f\"variant store {var_store['name']} FAILED:\")\n",
    "    print(e)\n",
    "\n",
    "var_store = omics.get_variant_store(name=var_store['name'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6bb0979",
   "metadata": {},
   "source": [
    "### Import VCF files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "73a348de",
   "metadata": {},
   "outputs": [],
   "source": [
    "l_vcf = [dict(zip([\"source\"],[uri])) for i, uri in enumerate(vcf_list)]\n",
    "\n",
    "response = omics.start_variant_import_job(destinationName=var_store['name'], \n",
    "                                          roleArn=get_role_arn(omics_iam_name),\n",
    "                                          items=l_vcf)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc6a56d1",
   "metadata": {},
   "source": [
    "## Query Variant Store with Amazon Athena"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7058ce3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# To run Athena queries on the data, use AWS LakeFormation to create resource links to the database\n",
    "# and ensure IAM user running this notebook is a Data Lake Administrator.\n",
    "\n",
    "ram = boto3.client('ram')\n",
    "glue = boto3.client('glue')\n",
    "\n",
    "caller_identity = boto3.client('sts').get_caller_identity()\n",
    "AWS_ACCOUNT_ID = caller_identity['Account']\n",
    "AWS_IDENITY_ARN = caller_identity['Arn']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06fe7107",
   "metadata": {},
   "outputs": [],
   "source": [
    "response = ram.list_resources(resourceOwner='OTHER-ACCOUNTS', resourceType='glue:Database')\n",
    "\n",
    "if not response.get('resources'):\n",
    "    print('no shared resources found. verify that you have successfully created an Omics Analytics store')\n",
    "else:\n",
    "    variantstore_resources = [resource for resource in response['resources'] if var_store['id'] in resource['arn']]\n",
    "    if not variantstore_resources:\n",
    "        print(f\"no shared resources matching variant store id {var_store['id']} found\")\n",
    "    else:\n",
    "        variantstore_resource = variantstore_resources[0]\n",
    "\n",
    "variantstore_resource\n",
    "\n",
    "resource_share = ram.get_resource_shares(\n",
    "    resourceOwner='OTHER-ACCOUNTS', \n",
    "    resourceShareArns=[variantstore_resource['resourceShareArn']])['resourceShares'][0]\n",
    "resource_share\n",
    "\n",
    "# this creates a resource link to the table for the variant store and adds it to the `omicsdb` database\n",
    "glue.create_table(\n",
    "    DatabaseName='omicsdb',\n",
    "    TableInput = {\n",
    "        \"Name\": var_store['name'],\n",
    "        \"TargetTable\": {\n",
    "            \"CatalogId\": resource_share['owningAccountId'],\n",
    "            \"DatabaseName\": f\"variant_{AWS_ACCOUNT_ID}_{var_store['id']}\",\n",
    "            \"Name\": var_store['name'],\n",
    "        }\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e570e5d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use Athena engine version 3 to query Omics Analytic Stores \n",
    "\n",
    "athena = boto3.client('athena')\n",
    "\n",
    "athena_workgroups = athena.list_work_groups()['WorkGroups']\n",
    "\n",
    "athena_workgroup = None\n",
    "for wg in athena_workgroups:\n",
    "    print(wg['EngineVersion']['EffectiveEngineVersion'])\n",
    "    if wg['EngineVersion']['EffectiveEngineVersion'] == 'Athena engine version 3':\n",
    "        print(f\"Workgroup '{wg['Name']}' found using Athena engine version 3\")\n",
    "        athena_workgroup = wg\n",
    "        break\n",
    "else:\n",
    "    print(\"No workgroups with Athena engine version 3 found. creating one\")\n",
    "    athena_workgroup = athena.create_work_group(\n",
    "        Name='omics',\n",
    "        Configuration={\n",
    "            \"EngineVersion\": {\n",
    "                \"SelectedEngineVersion\": \"Athena engine version 3\"\n",
    "            }\n",
    "        }\n",
    "    )\n",
    "\n",
    "athena_workgroup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "8dbf203d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleid</th>\n",
       "      <th>contigname</th>\n",
       "      <th>start</th>\n",
       "      <th>referenceallele</th>\n",
       "      <th>alternatealleles</th>\n",
       "      <th>calls</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3dcf722a-8e47-dd42-a0f3-97496f4e8d09</td>\n",
       "      <td>1</td>\n",
       "      <td>46932823</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>[0, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2a19fd41-040b-e9fb-fb5c-e9095f04922e</td>\n",
       "      <td>1</td>\n",
       "      <td>46932823</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>[0, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b3eaed13-18ab-509f-f9c7-2d0337181c7b</td>\n",
       "      <td>1</td>\n",
       "      <td>46932823</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>[0, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>08dc48f1-2321-362b-8c26-9ec0ed06f4fd</td>\n",
       "      <td>1</td>\n",
       "      <td>46932823</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>[0, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>a870a2c3-9fd9-7141-6391-b1428fb920bb</td>\n",
       "      <td>1</td>\n",
       "      <td>46932823</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>[0, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1acc5aa3-ba78-a290-7cf6-75284343ab22</td>\n",
       "      <td>1</td>\n",
       "      <td>46932823</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>[0, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1d08cfdb-2a91-55bd-4e27-2e88630849f3</td>\n",
       "      <td>1</td>\n",
       "      <td>46932823</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>[0, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>6df219b3-88cd-b76a-2c14-40a6399152aa</td>\n",
       "      <td>1</td>\n",
       "      <td>46932823</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>[0, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7a4dad5f-3b06-3104-8540-f20982d2c4c5</td>\n",
       "      <td>1</td>\n",
       "      <td>46932823</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>[0, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>30fe95e3-951f-ef8c-4b1d-c921a9a6dbdb</td>\n",
       "      <td>1</td>\n",
       "      <td>46932823</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>[0, 1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               sampleid contigname     start referenceallele  \\\n",
       "0  3dcf722a-8e47-dd42-a0f3-97496f4e8d09          1  46932823               A   \n",
       "1  2a19fd41-040b-e9fb-fb5c-e9095f04922e          1  46932823               A   \n",
       "2  b3eaed13-18ab-509f-f9c7-2d0337181c7b          1  46932823               A   \n",
       "3  08dc48f1-2321-362b-8c26-9ec0ed06f4fd          1  46932823               A   \n",
       "4  a870a2c3-9fd9-7141-6391-b1428fb920bb          1  46932823               A   \n",
       "5  1acc5aa3-ba78-a290-7cf6-75284343ab22          1  46932823               A   \n",
       "6  1d08cfdb-2a91-55bd-4e27-2e88630849f3          1  46932823               A   \n",
       "7  6df219b3-88cd-b76a-2c14-40a6399152aa          1  46932823               A   \n",
       "8  7a4dad5f-3b06-3104-8540-f20982d2c4c5          1  46932823               A   \n",
       "9  30fe95e3-951f-ef8c-4b1d-c921a9a6dbdb          1  46932823               A   \n",
       "\n",
       "  alternatealleles   calls  \n",
       "0              [G]  [0, 1]  \n",
       "1              [G]  [0, 1]  \n",
       "2              [G]  [0, 1]  \n",
       "3              [G]  [0, 1]  \n",
       "4              [G]  [0, 1]  \n",
       "5              [G]  [0, 1]  \n",
       "6              [G]  [0, 1]  \n",
       "7              [G]  [0, 1]  \n",
       "8              [G]  [0, 1]  \n",
       "9              [G]  [0, 1]  "
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Use AWS Wrangler to submit query and get results as a Pandas Dataframe\n",
    "\n",
    "import awswrangler as wr\n",
    "\n",
    "# Replace <variant-table> with the name of the annotation table created in the previous step\n",
    "\n",
    "df_var = wr.athena.read_sql_query(\"select sampleid, contigname, start, referenceallele, alternatealleles, calls from <variant-table> limit 10;\", \n",
    "                              database=\"omicsdb\", workgroup = \"omics\")\n",
    "df_var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef5d4866",
   "metadata": {},
   "source": [
    "## Create Annotation Store and import ClinVar annotation file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b747f19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace <S3 URI for annotation file> with the S3 URI of the ClinVar annotation file \n",
    "\n",
    "SOURCE_ANNOTATION_URI = \"<S3 URI for annotation file>\"\n",
    "\n",
    "ann_store_name = f'synthea_annotations_{ts.lower()}'\n",
    "\n",
    "# Replace <reference name> with reference genome name created in the previous steps\n",
    "ref_name = '<reference name>' \n",
    "\n",
    "response = omics.create_annotation_store(\n",
    "    name=ann_store_name, \n",
    "    reference={\"referenceArn\": get_reference_arn(ref_name, omics)},\n",
    "    storeFormat='VCF'\n",
    ")\n",
    "\n",
    "ann_store = response\n",
    "response\n",
    "\n",
    "try:\n",
    "    waiter = omics.get_waiter('annotation_store_created')\n",
    "    waiter.wait(name=ann_store['name'])\n",
    "\n",
    "    print(f\"annotation store {ann_store['name']} ready for use\")\n",
    "except botocore.exceptions.WaiterError as e:\n",
    "    print(f\"annotation store {ann_store['name']} FAILED:\")\n",
    "    print(e)\n",
    "\n",
    "ann_store = omics.get_annotation_store(name=ann_store['name'])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29db5e49",
   "metadata": {},
   "source": [
    "### !!! Wait until the Annotation Store is created !!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "534c2b53",
   "metadata": {},
   "outputs": [],
   "source": [
    "source = urllib.parse.urlparse(SOURCE_ANNOTATION_URI)\n",
    "bucket = source.netloc\n",
    "prefix = source.path[1:]\n",
    "\n",
    "s3r = boto3.resource('s3')\n",
    "\n",
    "bucket = s3r.Bucket(bucket)\n",
    "objects = bucket.objects.filter(Prefix=prefix, MaxKeys=10_000)\n",
    "ext = '.vcf'\n",
    "\n",
    "annotation_list = [f\"s3://{o.bucket_name}/{o.key}\" for o in objects if o.key.endswith(ext)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "167610f7",
   "metadata": {},
   "source": [
    "### Import annotation file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "330bff87",
   "metadata": {},
   "outputs": [],
   "source": [
    "response = omics.start_annotation_import_job(\n",
    "    destinationName=ann_store['name'],\n",
    "    roleArn=get_role_arn(omics_iam_name),\n",
    "    items=[{\"source\": SOURCE_ANNOTATION_URI}]\n",
    ")\n",
    "response"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b1199c0",
   "metadata": {},
   "source": [
    "## Query Annotation Store with Amazon Athena"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b614d7ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# To run Athena queries on the data, use AWS LakeFormation to create resource links to the database and \n",
    "\n",
    "response = ram.list_resources(resourceOwner='OTHER-ACCOUNTS', resourceType='glue:Database')\n",
    "\n",
    "if not response.get('resources'):\n",
    "    print('no shared resources found. verify that you have successfully created an Omics Analytics store')\n",
    "else:\n",
    "    annotationstore_resources = [resource for resource in response['resources'] if ann_store['id'] in resource['arn']]\n",
    "    if not annotationstore_resources:\n",
    "        print(f\"no shared resources matching annotation store id {ann_store['id']} found\")\n",
    "    else:\n",
    "        annotationstore_resource = annotationstore_resources[0]\n",
    "\n",
    "        resource_share = ram.get_resource_shares(\n",
    "            resourceOwner='OTHER-ACCOUNTS', \n",
    "            resourceShareArns=[annotationstore_resource['resourceShareArn']])['resourceShares'][0]\n",
    "        \n",
    "        # this creates a resource link to the table for the annotation store and adds it to the `omicsdb` database\n",
    "        glue.create_table(\n",
    "            DatabaseName='omicsdb',\n",
    "            TableInput = {\n",
    "                \"Name\": ann_store['name'],\n",
    "                \"TargetTable\": {\n",
    "                    \"CatalogId\": resource_share['owningAccountId'],\n",
    "                    \"DatabaseName\": f\"annotation_{AWS_ACCOUNT_ID}_{ann_store['id']}\",\n",
    "                    \"Name\": ann_store['name'],\n",
    "                }\n",
    "            }\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "b0f4bf17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contigname</th>\n",
       "      <th>start</th>\n",
       "      <th>referenceallele</th>\n",
       "      <th>alternatealleles</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>926009</td>\n",
       "      <td>G</td>\n",
       "      <td>[T]</td>\n",
       "      <td>[(CLNSIG, Likely_benign), (GENEINFO, SAMD11:14...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>926026</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>[(CLNSIG, Likely_benign), (GENEINFO, SAMD11:14...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>925975</td>\n",
       "      <td>T</td>\n",
       "      <td>[C]</td>\n",
       "      <td>[(CLNSIG, Uncertain_significance), (GENEINFO, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>926002</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>[(CLNSIG, Uncertain_significance), (GENEINFO, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>926013</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>[(CLNSIG, Uncertain_significance), (GENEINFO, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>926024</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>[(CLNSIG, Likely_benign), (GENEINFO, SAMD11:14...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>925955</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>[(CLNSIG, Likely_benign), (GENEINFO, SAMD11:14...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>925968</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>[(CLNSIG, Likely_benign), (GENEINFO, SAMD11:14...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>925985</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>[(CLNSIG, Likely_benign), (GENEINFO, SAMD11:14...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>925951</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>[(RS, 1640863258), (CLNSIG, Uncertain_signific...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  contigname   start referenceallele alternatealleles  \\\n",
       "0          1  926009               G              [T]   \n",
       "1          1  926026               C              [T]   \n",
       "2          1  925975               T              [C]   \n",
       "3          1  926002               C              [T]   \n",
       "4          1  926013               G              [A]   \n",
       "5          1  926024               G              [A]   \n",
       "6          1  925955               C              [T]   \n",
       "7          1  925968               C              [T]   \n",
       "8          1  925985               C              [T]   \n",
       "9          1  925951               G              [A]   \n",
       "\n",
       "                                          attributes  \n",
       "0  [(CLNSIG, Likely_benign), (GENEINFO, SAMD11:14...  \n",
       "1  [(CLNSIG, Likely_benign), (GENEINFO, SAMD11:14...  \n",
       "2  [(CLNSIG, Uncertain_significance), (GENEINFO, ...  \n",
       "3  [(CLNSIG, Uncertain_significance), (GENEINFO, ...  \n",
       "4  [(CLNSIG, Uncertain_significance), (GENEINFO, ...  \n",
       "5  [(CLNSIG, Likely_benign), (GENEINFO, SAMD11:14...  \n",
       "6  [(CLNSIG, Likely_benign), (GENEINFO, SAMD11:14...  \n",
       "7  [(CLNSIG, Likely_benign), (GENEINFO, SAMD11:14...  \n",
       "8  [(CLNSIG, Likely_benign), (GENEINFO, SAMD11:14...  \n",
       "9  [(RS, 1640863258), (CLNSIG, Uncertain_signific...  "
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Replace <annotation-table> with the name of the annotation table created in the previous step\n",
    "\n",
    "df_ann = wr.athena.read_sql_query(\"select contigname, start, referenceallele, alternatealleles, attributes from <annotation-table> order by contigname limit 10;\", \n",
    "                              database=\"omicsdb\", workgroup = \"omics\")\n",
    "df_ann"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77f9922d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "065dc264",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda_python3",
   "language": "python",
   "name": "conda_python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
